% MATERIALS AND METHODS 
% NOTE: be careful to escape special characters like #.

\subsection{Animals}

We generated Washc4 mutant (SWIP\textsuperscript{P1019R}) mice in collaboration
with the Duke Transgenic Core Facility to mimic the de novo human variant at
amino acid 1019 of human WASHC4. A CRISPR-induced CCT>CGT point mutation was
introduced into exon 29 of Washc4. 50ng/µl of the sgRNA
(5'-ttgagaatactcacaagaggagg-3'), 100ng/µl Cas9 mRNA, and 100ng/µl of a repair
oligonucleotide containing the C>G mutation were injected into the cytoplasm of
B6SJLF1/J mouse embryos (Jax \#100012) (See Table S4 for the sequence of the
repair oligonucleotide). Mice with germline transmission were then backcrossed
into a C57BL/ 6J background (Jax \#000664). At least 5 backcrosses were obtained
before animals were used for behavior. We bred heterozygous
SWIP\textsuperscript{P1019R} mice together to obtain age-matched mutant and
wild-type genotypes for cell culture and behavioral experiments.  Genetic
sequencing was used to screen for germline transmission of the C>G point
mutation (FOR: 5'-tgcttgtagatgtttttcct-3', REV: 5'-gttaacatgatcctatggcg-3'). All
mice were housed in the Duke University's Division of Laboratory Animal
Resources or Behavioral Core facilities at 2-5 animals/cage on a 14:10h
light:dark cycle. All experiments were conducted with a protocol approved by the
Duke University Institutional Animal Care and Use Committee in accordance with
NIH guidelines. 

\subsection{Human Subjects}

We retrospectively analyzed clinical findings from seven children with
homozygous WASHC4c.3056C>G mutations (obtained by Dr. Rajab in 2010 at the Royal
Hospital, Muscat, Oman). The original report of these human subjects and
parental consent for data use can be found in (Ropers et al., 2011). 

\subsection{Cell Lines}

HEK293T cells (ATCC \#CRL-11268) were purchased from the Duke Cell Culture
facility. and were tested for mycoplasma contamination. HEK239T cells were used
for co-immunoprecipitation experiments and preparation of AAV viruses.

\subsection{Primary Neuronal Culture}

Primary neuronal cultures were prepared from P0 mouse cortex. P0 mouse pups were
rapidly decapitated and cortices were dissected and kept individually in 5ml
Hibernate A (Thermo \#A1247501) supplemented with 2\% B27(Thermo \#17504044) at
4ºC overnight to allow for individual animal genotyping before plating. Neurons
were then treated with Papain (Worthington \#LS003120) and DNAse (VWR
\#V0335)-supplemented Hibernate A for 18min at 37ºC and washed twice with
plating medium (plating medium: Neurobasal A (Thermo \#10888022) supplemented
with 10\% horse serum, 2\% B-27, and 1\% GlutaMAX (Thermo \#35050061)), and
triturated before plating at 250,000 cells/well on poly-L-lysine-treated
coverslips (Sigma \#P2636) in 24-well plates. Plating medium was replaced with
growth medium (Neurobasal A, 2\% B-27, 1\% GlutaMAX) 2 hours later. Cell media
was supplemented and treated with AraC at DIV5 (5uM final concentration/well).
Half-media changes were then performed every 4 days. 

\subsection{Plasmid DNA Constructs}

For immunoprecipitation experiments, a pmCAG-SWIP-WT-HA construct was generated
by PCR amplification of the human WASHC4 sequence, which was then inserted
between NheI and SalI restriction sites of a pmCAG-HA backbone generated in our
lab. Site-directed mutagenesis (Agilent \#200517) was used to introduce a C>G
point mutation into this pmCAG-SWIP-WT-HA construct for generation of a
pmCAG-SWIP-MUT-HA construct (FOR:
5'-ctacaaagttgagggtcagacggggaacaattatatagaaa-3', REV:
5'-tttctatataattgttccccgtctgaccctcaactttgtag-3’). For iBioID experiments, an AAV
construct expressing hSyn1-WASH1-BioID2-HA was generated by cloning a Washc1
insert between SalI and HindIII sites of a pAAV-hSyn1-Actin
Chromobody-Linker-BioID2-pA construct (replacing Actin Chromobody) generated in
our lab. This backbone included a 25nm GS linker-BioID2-HA fragment from Addgene
\#80899, generated by Kim et al. (Kim et al., 2016). An hSyn1-solubleBioID2-HA
construct was created similarly, by removing Actin Chromobody from the above
construct. Oligonucleotide sequences are reported in Table S4. Sequences of the
plasmid DNA constructs are available online (see Key Resources Table). 

\subsection{AAV Viral Preparation}

AAV preparations were performed as described previously(Uezu et al., 2016). The
day before transfection, HEK293T cells were plated at a density of 1.5x107 cells
per 15cm2 plate in DMEM media with 10\% fetal bovine serum and 1\% Pen/Strep
(Thermo \#11965-092, Sigma \#F4135, Thermo \#15140-122). Six HEK293T 15cm2
plates were used per viral preparation. The next day, 30µg of pAd-DeltaF6 helper
plasmid, 15µg of AAV2/9 plasmid, and 15µg of an AAV plasmid carrying the
transgene of interest were mixed in OptiMEM with PEI-MAX (final concentration
80µg/ml, Polysciences \#24765). 2ml of this solution were then added dropwise to
each of the 6 HEK293T 15cm2 plates. Eight hours later, the media was replaced
with 20ml DMEM+10\%FBS. 72 hours post-transfection, cells were scraped and
collected in the media, pooled, and centrifuged at 1,500rpm for 5min at RT. The
final pellet from the 6 cell plates was resuspended in 5ml of cell lysis buffer
(15 mM NaCl, 5 mM Tris-HCl, pH 8.5), and freeze-thawed three times using an
ethanol/dry ice bath. The lysate was then treated with 50U/ml of Benzonase
(Novagen \#70664), for 30min in a 37ºC water bath, vortexed, and then
centrifuged at 4,500rpm for 30min at 4ºC. The resulting supernatant containing
AAV particles was added to the top of an iodixanol gradient (15\%, 25\%, 40\%,
60\% top to bottom) in an Optiseal tube (Beckman Coulter \#361625). The gradient
was then centrifuged using a Beckman Ti-70 rotor in a Beckman XL-90
ultracentrifuge at 67,000rpm for 70min, 18ºC. The purified viral solution was
extracted from the 40\%/60\% iodixanol interface using a syringe, and placed
into an Amicon 100kDa filter unit \#UFC910024). The viral solution was washed in
this filter 3 times with 1X ice-cold PBS by adding 5ml of PBS and centrifuging
at 4,900rpm for 45min at 4ºC to obtain a final volume of approximately 200µl of
concentrated virus that was aliquoted into 5-10µl aliquots and stored at -80ºC
until use. 

\subsection{Immunocytochemistry}

At DIV15, neurons were fixed for 15 minutes using ice-cold 4\%PFA/4\% sucrose in
1X PBS, pH 7.4 (for EEA1 staining), or 30 minutes with 50\% Bouin’s solution/4\%
sucrose (for CathepsinD staining, Sigma \#HT10132), pH 7.4(Cheng et al., 2018).
Fixed neurons were washed with 1X PBS, then permeabilized with 0.25\%
TritonX-100 in PBS for 8 minutes at room temperature (RT), and blocked with
5\%normal goat serum/0.2\%Triton-X100 in PBS (blocking buffer) for 1 hour at RT
with gentle rocking. For EEA1/MAP2 staining, samples were incubated with primary
antibodies diluted in blocking buffer at RT for 1 hour. For CathepsinD/MAP2
staining, samples were incubated with primary antibodies diluted in blocking
buffer overnight at 4ºC. For both conditions, samples were washed three times
with 1X PBS, and incubated for 30min at RT with secondary antibodies, protected
from light. After secondary antibody staining, coverslips were washed three
times with 1X PBS, and mounted with FluoroSave mounting solution (Sigma
\#345789). See antibody section for primary and secondary antibody
concentrations. 

\subsubsection{Primary antibodies} Rabbit anti-EEA1 (Cell Signaling
Technology \#C45B10, 1:500), Rat anti-CathepsinD (Novus \#204712, 1:250), Guinea
Pig anti-MAP2 (Synaptic Systems \#188004, 1:500)

\subsubsection{Secondary antibodies} Goat anti-Rabbit Alexa Fluor 568
(Invitrogen \#A11036, 1:1000), Goat anti-Guinea Pig Alexa Fluor 488 (Invitrogen
\#A11073, 1:1000), Goat anti-Rat Alexa Fluor 488 (Invitrogen \#A11006, 1:1000),
Goat anti-Guinea Pig Alexa Fluor 555 (Invitrogen \#A21435, 1:1000)

\subsection{Immunohistochemistry}

Mice were deeply anesthetized with isoflurane and then transcardially perfused
with ice-cold heparinized PBS (25U/ml) by gravity flow. After clearing of liver
and lungs (~2min), perfusate was switched to ice-cold 4\% PFA in 1X PBS (pH 7.4)
for 15 minutes. Brains were dissected, post-fixed in 4\%PFA overnight at 4ºC,
and then cryoprotected in 30\% sucrose/1X PBS for 48hr at 4ºC. Fixed brains were
then mounted in OTC (Sakura TissueTek \#4583) and stored at -20ºC until
cryosectioning. Every third sagittal section (30 µm thickness) was collected
from the motor cortex and striatal regions. Free-floating sections were then
permeabilized with 1\%TritonX-100 in 1X PBS at RT for 2 hr, and blocked in 1X
blocking solution (Abcam \#126587) diluted in 0.2\%TritonX-100 in 1X PBS for 1hr
at RT. Sections were then incubated in primary antibodies diluted in the 1X
blocking solution for two overnights at 4ºC. After three washes with
0.2\%TritonX-100 in 1X PBS, the sections were then incubated in secondary
antibodies diluted in 1X blocking buffer for one overnight at 4ºC. Sections were
then washed four times with 0.2\%TritonX-100 in 1X PBS at RT, and mounted onto
coverslips with FluoroSave mounting solution (Sigma \#345789). 

\subsubsection{Primary antibodies} Rabbit anti-Cleaved Caspase-3 (Cell
Signaling Technology \#9661, 1:2000), Mouse anti-Calbindin (Sigma \#C9848,
1:2000), Rat anti-HA 3F10 (Sigma \#12158167001, 1:500)

\subsubsection{Secondary antibodies} Donkey anti-Rabbit Alexa Fluor 488
(Invitrogen \#A21206, 1:2000), Goat anti-Mouse Alexa Fluor 594 (Invitrogen
\#A11032, 1:2000), Goat anti-Rat Alexa Fluor 488 (Invitrogen \#A11006, 1:5000),
Streptavidin Alexa Fluor 594 conjugate (Invitrogen \#S32356, 1:5000),
4',6-diamidino-2-phenylindole (DAPI, Sigma \#D9542, 1:1000 for 10min at RT)

\subsection{Western Blotting}

Ten micrograms of each sample were electrophoresed through a 12-well, 4-20\%
SDS-PAGE gel (Bio-Rad \#4561096) at 100V for 1hr at RT, transferred onto a
nitrocellulose membrane (GE Life Sciences \#GE10600002) at 100V for 70min at RT
on ice, and blocked with 5\% nonfat dry milk in TRIS-buffered saline containing
0.05\% Tween-20 (TBST, pH 7.4). Gels were saved for Coomassie staining at RT for
30 min. Membranes were probed with one primary antibody at a time for 24hr at
4ºC, then washed four times with TBST at RT before incubating with the
corresponding species-specific secondary antibody at RT for 1hr. Membranes were
washed with TBST, and then enhanced chemiluminescence (ECL) substrate was added
(Thermo Fischer \#32109). Membranes were exposed to autoradiography films and
scanned with an Epson 1670 at 600dpi. We probed each membrane with one antibody
at a time, then stripped the membrane with stripping buffer (Thermo Fischer
\#21059) for 10min at RT, and then blocked for 1hr at RT before probing with the
next antibody. Order of probes: Strumpellin, then \textbeta-Tubulin, then WASH1.
We determined the optical density of the bands using Image J software (NIH).
Data obtained from three independent experiments were plotted and statistically
analyzed using GraphPad Prism (version 8) software.

\subsubsection{Primary antibodies} Rabbit anti-Strumpellin
(Santa Cruz \#sc-87442, 1:500), Rabbit anti-WASH1 c-terminal (Sigma
\#SAB4200373, 1:500), Mouse anti-Beta Tubulin III (Sigma \#T8660, 1:10,000),
Mouse anti-HA (BioLegend \#MMS-101P, 1:5000)

\subsubsection{Secondary antibodies} Donkey anti-Rabbit-HRP
(GE Life Sciences \#NA934, 1:5,000), Goat anti-mouse-HRP (GE Life Sciences
\#NA931, 1:5000)

\subsection{Immunoprecipitation}

HEK293T cells were transfected with pmCAG-SWIP-WT-HA or pmCAG-SWIP-MUT-HA
constructs for three days, as previously described(Mason et al., 2011). Cells
were lysed with lysis buffer (25mM HEPES, 150mM NaCl, 1mM EDTA, 1\% NonidetP-40,
pH 7.4) containing protease inhibitors (5mM NaF, 1mM orthovanadate, 1mM AEBSF,
and 2 μg/mL leupeptin/pepstatin) and centrifuged at 1,700g for 5 min. Collected
supernatant was incubated with 30µl of pre-washed anti-HA agarose beads (Sigma
\#A2095) on a sample rotator (15 rpm) for 2 hrs at 4ºC. Beads were then washed 3
times with lysis buffer, and sample buffer was added before subjecting to
immunoblotting as described above. The protein-transferred membrane was probed
individually for WASH1, Strumpellin, and HA. Data were collected from four
separate preparations of WT and MUT conditions. 

\subsection{Electron Microscopy}

Adult (7mo) WT and MUT SWIP\textsuperscript{P1019R} mice were deeply
anesthetized with isoflurane and then transcardially perfused with warmed
heparinized saline (25U/ml heparin) for 4 minutes, followed by ice-cold 0.15M
cacodylate buffer pH 7.4 containing 2.5\% glutaraldehyde (Electron Microscopy
Sciences \#16320), 3\% paraformaldehyde, and 2mM CaCl2 for 15 minutes. Brain
samples were dissected and stored on ice in the same fixative for 2 hours before
washing in 0.1M sodium cacodylate buffer (3 changes for 15 minutes each).
Samples were then post-fixed in 1.0\% OsO4 in 0.1 M Sodium cacodylate buffer for
1 hour on a rotator. Samples were then washed in 3, 15-minute changes of 0.1M
sodium cacodylate. Samples were then placed into en bloc stain (1\% uranyl
acetate) overnight at 4\celsius. Subsequently, samples were dehydrated in a
series of ascending acetone concentrations including 50\%, 70\%, 95\%, and 100\%
for three cycles with 15 minutes incubation at each concentration change.
Samples were then placed in a 50:50 mixture of epoxy resin (Epon) and acetone
overnight on a rotator. This solution was then replaced twice with 100\% fresh
Epon for at least 2 hours at room temperature on a rotator. Samples were
embedded with 100\% Epon resin in BEEM capsules (Ted Pella) for 48 hours at
60\celsius.

Samples were ultrathin sectioned to 60-70nm on a Reichert
Ultracut E ultramicrotome. Harvested grids were then stained with 2\% uranyl
acetate in 50\% ethanol for 30 minutes and Sato’s lead stain for 1 min.
Micrographs were acquired using a Phillips CM12 electron microscope operating at
80Kv, at 1700x magnification. Micrographs were analyzed in Adobe Photoshop 2019,
using the “magic wand” tool to demarcate and measure the area of electron-dense
and electron-lucent regions of interest (ROIs). Statistical analyses of ROI
measurements were performed in GraphPad Prism (version 8) software. The
experimenter was blinded to genotype for image acquisition and analysis. 

\subsection{iBioID Sample Preparation}

AAV2/9 viral probes, hSyn1-WASH1-BioID2-HA or hSyn1-solubleBioID2-HA, were
injected into wild-type CD1 mouse brains using a Hamilton syringe (\#7635-01) at
age P0-P1 to ensure viral spread throughout the forebrain(Glascock et al.,
2011). 15 days post-viral injection, biotin was subcutaneously administered at
24mg/kg for seven consecutive days for biotinylation of proteins in proximity to
BioID2 probes. Whole brains were extracted on the final day of biotin
injections, snap frozen, and stored in liquid nitrogen until protein
purification. Seven brains were used for protein purification of each probe, and
each purification was performed three times independently (21 brains total for
WASH1-BioID2, 21 for solubleBioID2).

We performed all homogenization and protein purification on ice. A 2ml Dounce
homogenizer was used to individually homogenize each brain in a 1:1 solution of
Lysis-R:2X-RIPA buffer solution with protease inhibitors (Roche cOmplete tablets
\#11836153001). Each sample was sonicated three times for 7 seconds and then
centrifuged at 5000g for 5min at 4ºC. Samples were transferred to Beckman
Coulter 1.5ml tubes (\#344059), and then spun at 45,000rpm in a Beckman Coulter
tabletop ultracentrifuge (TLA-55 rotor) for 1hr at 4ºC. SDS was added to
supernatants (final 1\%) and samples were then boiled for 5min at 95ºC. We next
combined supernatants from the same condition together (WASH1-BioID2 vs.
solubleBioID2) in 15ml conical tubes to rotate with 30µl high-capacity
NeutrAvidin beads overnight at 4ºC (Thermo \#29204).

The following day, all steps were performed under a hood with keratin-free
reagents. Samples were spun down at 6000rpm, 4ºC for 5min to pellet the beads
and remove supernatant. The pelleted beads then went through a series of washes,
each for 10 min at RT with 500ul of solvent, and then spun down on a tabletop
centrifuge to pellet the beads for the next wash. The washes were as follows:
2\% SDS twice, 1\% TritonX100-1\%deoxycholate-25mM LiCl2 once, 1M NaCL twice,
50mM Ammonium Bicarbonate (Ambic) five times. Beads were then mixed 1:1 with a
2X Laemmli sample buffer that contained 3mM biotin/50mM Ambic, boiled for 5 mins
at 95ºC, vortexed three times, and then biotinylated protein supernatants were
stored at -80ºC until LC-MS/MS. 

\subsection{LC-MS/MS for iBioID} We gave the Duke Proteomics and Metabolomics
Shared Resource (DPMSR) six eluents from streptavidin resins (3 x WASH1-BioID2,
3 x solubleBioID2), stored on dry ice. Samples were reduced with 10 mM
dithiolthreitol for 30 min at 80ºC and alkylated with 20 mM iodoacetamide for 30
min at room temperature. Next, samples were supplemented with a final
concentration of 1.2\% phosphoric acid and 256 μL of S-Trap (Protifi) binding
buffer (90\% MeOH/100mM TEAB). Proteins were trapped on the S-Trap, digested
using 20 ng/μl sequencing grade trypsin (Promega) for 1 hr at 47ºC, and eluted
using 50 mM TEAB, followed by 0.2\% FA, and lastly using 50\% ACN/0.2\% FA. All
samples were then lyophilized to dryness and resuspended in 20 μL 1\%TFA/2\%
acetonitrile containing 25 fmol/μL yeast alcohol dehydrogenase (UniProtKB
P00330; ADH\_YEAST). From each sample, 3 μL was removed to create a pooled QC
sample (SPQC) which was run analyzed in technical triplicate throughout the
acquisition period.

Quantitative LC/MS/MS was performed on 2 μL of each sample, using a nanoAcquity
UPLC system (Waters) coupled to a Thermo QExactive HF-X high resolution accurate
mass tandem mass spectrometer (Thermo) via a nanoelectrospray ionization source.
Briefly, the sample was first trapped on a Symmetry C18 20 mm × 180 μm trapping
column (5 μl/min at 99.9/0.1 v/v water/acetonitrile), after which the analytical
separation was performed using a 1.8 μm Acquity HSS T3 C18 75 μm × 250 mm column
(Waters) with a 90-min linear gradient of 5 to 30\% acetonitrile with 0.1\%
formic acid at a flow rate of 400 nanoliters/minute (nL/min) with a column
temperature of 55ºC. Data collection on the QExactive HF-X mass spectrometer was
performed in a data-dependent acquisition (DDA) mode of acquisition with a
r=120,000 (@ m/z 200) full MS scan from m/z 375 – 1600 with a target AGC value
of 3e6 ions followed by 30 MS/MS scans at r=15,000 (@ m/z 200) at a target AGC
value of 5e4 ions and 45 ms. A 20s dynamic exclusion was employed to increase
depth of coverage. The total analysis cycle time for each sample injection was
approximately 2 hours. 

\subsection{LOPIT-DC Subcellular Fractionation}

We performed three independent fractionation experiments with one adult SWIP
mutant brain and one WT mouse brain fractionated in each experiment. Each mouse
was sacrificed by isoflurane inhalation and its brain was immediately extracted
and placed into a 2ml Dounce homogenizer on ice with 1ml isotonic TEVP
homogenization buffer (320mM sucrose, 10mM Tris base, 1mM EDTA, 1mM EGTA, 5mM
NaF, pH7.4 (Hallett et al., 2008)). A cOmplete mini protease inhibitor cocktail
tablet (Sigma \#11836170001) was added to a 50ml TEVP buffer aliquot immediately
before use. Brains were homogenized for 15 passes with a Dounce homogenizer to
break the tissue, and then this lysate was brought up to a 5ml volume with
additional TEVP buffer. Lysates were then passed through a 0.5ml ball-bearing
homogenizer for two passes (14 µm ball, Isobiotec) to release organelles. Final
brain lysate volumes were approximately 7.5ml each. Lysates were then divided
into replicate microfuge tubes (Beckman Coulter \#357448) to perform
differential centrifugation, following Geladaki et. al’s LOPIT-DC
protocol(Geladaki et al., 2019). Centrifugation was carried out at 4ºC in a
tabletop Eppendorf 5424 centrifuge for spins at 200g, 1,000g, 3,000g, 5,000g,
9,000g, 12,000g, and 15,000g. To isolate the final three fractions, a tabletop
Beckman TLA-100 ultracentrifuge with a TLA-55 rotor was used at 4ºC with speeds
of: 30,000g, 79,000g, and 120,000g, respectively. Samples were kept on ice at
all times and pellets were stored at -80ºC. Pellets from seven fractions
(5,000g-120,000g) were used for proteomic analyses. 

\subsection{16-plex TMT LC-MS/MS} The Duke Proteomics and Metabolomics Shared
Resource (DPMSR) processed and prepared fraction pellets from all 42 frozen
samples simultaneously (7 fractions per brain from 3 WT and 3 MUT brains). Due
to volume constraints, each sample was split into 3 tubes, for a total of 126
samples, which were processed in the following manner: 100µL of 8M Urea was
added to the first aliquot then probe sonicated for 5 seconds with an energy
setting of 30\%. This volume was then transferred to the second and then third
aliquot after sonication in the same manner. All tubes were centrifuged at
10,000g and any residual volume from tubes 1 and 2 were added to tube 3. Protein
concentrations were determined by BCA on the supernatant in duplicate (5 μL each
assay). Total protein concentrations for each replicate ranged from 1.1 mg/mL to
7.8 mg/mL with total protein quantities ranging from 108.3 to 740.81 µg. 60 µg
of each sample was removed and normalized to 52.6µL with 8M Urea and 14.6µL 20\%
SDS. Samples were reduced with 10 mM dithiolthreitol for 30 min at 80ºC and
alkylated with 20 mM iodoacetamide for 30 min at room temperature. Next, they
were supplemented with 7.4 μL of 12\% phosphoric acid, and 574 μL of S-Trap
(Protifi) binding buffer (90\% MeOH/100mM TEAB). Proteins were trapped on the
S-Trap, digested using 20 ng/μl sequencing grade trypsin (Promega) for 1 hr at
47ºC, and eluted using 50 mM TEAB, followed by 0.2\% FA, and lastly using 50\%
ACN/0.2\% FA. All samples were then lyophilized to dryness.

Each sample was resuspended in 120 μL 200 mM triethylammonium bicarbonate, pH
8.0 (TEAB). From each sample, 20µL was removed and combined to form a pooled
quality control sample (SPQC). Fresh TMTPro reagent (0.5 mg for each 16-plex
reagent) was resuspended in 20 μL 100\% acetonitrile (ACN) and was added to each
sample. Samples were incubated for 1 hour at RT. After the 1-hour reaction, 5 μL
of 5\% hydroxylamine was added and incubated for 15 minutes at room temperature
to quench the reaction. Each 16-plex TMT experiment consisted of the WT and MUT
fractions from one mouse, as well as the 2 SPQC samples. Samples corresponding
to each experiment were concatenated and lyophilized to dryness.

Samples were resuspended in 800µL 0.1\% formic acid. 400µg was fractionated into
48 unique high pH reversed-phase fractions using pH 9.0 20 mM Ammonium formate
as mobile phase A and neat acetonitrile as mobile phase B. The column used was a
2.1 mm x 50 mm XBridge C18 (Waters) and fractionation was performed on an
Agilent 1100 HPLC with G1364C fraction collector. Throughout the method, the
flow rate was 0.4 mL/min and the column temperature was 55ºC. The gradient
method was set as follows: 0 min, 3\%B; 1 min, 7\% B; 50 min, 50\%B; 51 min,
90\% B; 55 min, 90\% B; 56 min, 3\% B; 70 min, 3\% B. 48 fractions were
collected in equal time segments from 0 to 52 minutes, then concatenated into 12
unique samples using every 12th fraction. For instance, fraction 1, 13, 25, and
37 were combined, fraction 2, 14, 26, and 38 were combined, etc. Fractions were
frozen and lyophilized overnight. Samples were resuspended in 66 μL 1\%TFA/2\%
acetonitrile prior to LC-MS analysis.

Quantitative LC/MS/MS was performed on 2 μL (1 μg) of each sample, using a
nanoAcquity UPLC system (Waters) coupled to a Thermo Orbitrap Fusion Lumos high
resolution accurate mass tandem mass spectrometer (Thermo) equipped with a FAIMS
Pro ion-mobility device via a nanoelectrospray ionization source. Briefly, the
sample was first trapped on a Symmetry C18 20 mm × 180 μm trapping column (5
μl/min at 99.9/0.1 v/v water/acetonitrile), after which the analytical
separation was performed using a 1.8 μm Acquity HSS T3 C18 75 μm × 250 mm column
(Waters) with a 90-min linear gradient of 5 to 30\% acetonitrile with 0.1\%
formic acid at a flow rate of 400 nanoliters/minute (nL/min) with a column
temperature of 55ºC. Data collection on the Fusion Lumos mass spectrometer was
performed for three different compensation voltages (CV: -40v, -60v, -80v).
Within each CV, a data-dependent acquisition (DDA) mode of acquisition with a
r=120,000 (\@ m/z 200) full MS scan from m/z 375 – 1600 with a target AGC value
of 4e5 ions was performed. MS/MS scans were acquired in the Orbitrap at r=50,000
(\@ m/z 200) from m/z 100 with a target AGC value of 1e5 and max fill time of 105
ms. The total cycle time for each CV was 1s, with total cycle times of 3 sec
between like full MS scans. A 45s dynamic exclusion was employed to increase
depth of coverage. The total analysis cycle time for each sample injection was
approximately 2 hours.

Following UPLC-MS/MS analyses, data were imported into Proteome Discoverer 2.4
(Thermo Scientific). The MS/MS data were searched against a SwissProt Mouse
database (downloaded November 2019) plus additional common contaminant proteins,
including yeast alcohol dehydrogenase (ADH), bovine casein, bovine serum
albumin, as well as an equal number of reversed-sequence “decoys” for FDR
determination. Mascot Distiller and Mascot Server (v 2.5, Matrix Sciences) were
utilized to produce fragment ion spectra and to perform the database searches.
Database search parameters included fixed modification on Cys (carbamidomethyl)
and variable modification on Met (oxidation), Asn/Gln (deamindation), Lys
(TMTPro) and peptide N-termini (TMTPro). Data were searched at 5 ppm precursor
and 0.02 product mass accuracy with full trypsin enzyme rules. Reporter ion
intensities were calculated using the Reporter Ions Quantifier algorithm in
Proteome Discoverer. Percolator node in Proteome Discoverer was used to annotate
the data at a maximum 1\% protein FDR.

\subsection{Mouse Behavioral Assays}

Behavioral tests were performed on age-matched WT and homozygous
SWIP\textsuperscript{P1019R} mutant littermates. Male and female mice were used
in all experiments. Testing was performed at two time points: P42-55 days old as
a young adult age, and 5.5 months old as mid-adulthood, so that we could compare
disease progression in this mouse model to human patients(Ropers et al., 2011).

The sequence of behavioral testing was: Y-maze (to measure working memory),
object novelty recognition (to measure short-term and long-term object
recognition memory), TreadScan (to assess gait), and steady-speed rotarod (to
assess motor control and strength) for 40-55 day old mice. Testing was performed
over 1.5 weeks, interspersed with rest days for acclimation. This sequence was
repeated with the same cohort at 5.5-6 months old, with three additional
measures added to the end of testing: fear conditioning (to assess associative
fear memory), a hearing test (to measure tone response), and a shock threshold
test (to assess somatosensation). Of note, a separate, second cohort of mice was
evaluated for fear conditioning, hearing, and shock threshold testing at
adolescence. After each trial, equipment was cleaned with Labsan to remove
residual odors. The experimenter was blinded to genotype for all behavioral
analyses.

\subsubsection{Y-maze}
Working memory was evaluated by measuring spontaneous
alternations in a 3-arm Y-maze under indirect illumination (80-90 lux). A mouse
was placed in the center of the maze and allowed to freely explore all arms,
each of which had different visual cues for spatial recognition. Trials were 5
min in length, with video data and analyses captured by EthoVision XT 11.0
software (Noldus Information Technology). Entry to an arm was define as the
mouse being >1 body length into a given arm. An alternation was defined as three
successive entries into each of the different arms. Total \% alternation was
calculated as the total number of alternations/the total number of arm entries
minus 2 x100.  

\subsubsection{Novel Object Recognition}
One hour before testing, mice were individually exposed to the testing arena (a 48 x 22 x 18cm white opaque arena)
for 10min under 80-100lux illumination without any objects. The test consisted
of three phases: training (day 1), short-term memory test (STM, day 1), and
long-term memory test (LTM, day 2). For the training phase, two identical
objects were placed 10 cm apart, against opposing walls of the arena. A mouse
was placed in the center of the arena and given full access to explore both
objects for 5 min and then returned to its home cage. For STM testing, one of
the training objects remained (the now familiar object), and a novel object
replaced one of the training objects (similar in size, different shape). The
mouse was returned to the arena 30 minutes after the training task and allowed
to explore freely for 5 mins. For LTM testing, the novel object was replaced
with another object, and the familiar object remained unchanged. The LTM test
was also 5 min in duration, conducted 24hr after the training task. Behavior was
scored using Ethovision 11.0 XT software (Noldus) and analyzed by a blind
observer. Object contact was defined as the mouse’s nose within 1 cm of the
object. We analyzed both number of nose contacts with each object and duration
of contacts. Preference scores were calculated as (duration contactnovel -
duration contactfamiliar) / total duration contactnovel+familiar. Positive
scores signified a preference for the novel object; whereas, negative scores
denoted a preference for the familiar object, and scores approaching zero
indicated no preference.  

\subsubsection{TreadScan}
A TreadScan forced locomotion treadmill
system (CleversSys Inc, Reston, Virginia) was used for gait recording and
analysis. Each mouse was recorded walking on a transparent treadmill at 45 days
old, and again at 5.5 months old.  Mice were acclimated to the treadmill chamber
for 1 minute before the start of recording to eliminate exploratory behavior
confounding normal gait. Trials were 20 seconds in length, with mice walking at
speeds between 13.83 and 16.53 cm/sec (P45 WT average 15.74 cm/s; P45 MUT
average 15.80 cm/s; 5.5mo WT average 15.77 cm/s; 5.5mo MUT average 15.85 cm/s).
A high-speed digital camera attached to the treadmill captured limb movement at
a frame rate of 100 frames/second. We used TreadScan software (CleversSys) and
representative WT and MUT videos to generate footprint templates, which were
then used to identify individual paw profiles for each limb. Parameters such as
stance time, swing time, step length, track width, and limb coupling were
recorded for the entire 20 sec duration for each animal. Output gait tracking
was verified manually by a blinded experimenter to ensure consistent limb
tracking throughout the duration of each video.  Steady Speed Rotarod A 5-lane
rotarod (Med Associates, St. Albans, VT) was used for steady-speed motor
analysis. The rod was run at a steady speed of 32rpm for four, 5-minute trials,
with a 40-minute inter-trial interval. We recorded mouse latency to fall by
infrared beam break, or manually for any mouse that completed two or more
rotations on the rod without walking. Mice were randomized across lanes for each
trial.

\subsubsection{Fear Conditioning}
Animals were examined in contextual and cued fear
conditioning as described by Rodriguiz and Wetsel(Rodriguiz and Wetsel, 2006).
Two separate cohorts of mice were used testing the two age groups. A three-day
testing paradigm was used to assess memory: conditioning on day 1, context
testing 24-hr post-conditioning on day 2, and cued tone testing 48hr
post-conditioning on day 3. All testing was conducted in fear conditioning
chambers (Med Associates). In the conditioning phase, mice were first acclimated
to the test chamber for two minutes under ~100 lux illumination. Then a 2900Hz,
80dB tone (conditioned stimulus, CS) played for 30 sec, which terminated with a
paired 0.4mA, 2 sec scrambled foot shock (unconditioned stimulus, US). Mice were
removed from the chamber and returned to their home cage 30 sec later. In the
context testing phase, mice were placed in the same conditioning chamber and
monitored for freezing behavior for a 5 min trial period, in the absence of the
CS and US. For cued tone testing, the chambers were modified to different
dimensions and shapes, contained different floor and wall textures, and lighting
was adjusted to 50 lux. Mice acclimated to the chamber for 2 min, and then the
CS was presented continuously for 3 min.  Contextual and cued fear memory was
assessed by freezing behavior, captured by automated video software
(CleversSys).

\subsubsection{Hearing Test}
We tested mouse hearing using a startle platform
(Med Associates) connected to Startle Pro Software in a sound-proof chamber.
Mice were placed in a ventilated restraint cylinder connected to the startle
response detection system to measure startle to each acoustic stimulus. After
two minutes of acclimation, mice were assessed for an acoustic startle response
to seven different tone frequencies, 2kHz, 3kHz, 4kHz, 8kHz, 12kHz, 16kHz, and
20kHz that were randomly presented three times each at four different decibels,
80, 100, 105, and 110dB, for a total of 84 trials. A random inter-trial interval
of 15-60 seconds (average 30sec) was used to prevent anticipation of a stimulus.
An animal’s reaction to the tone was recorded as startle reactivity in the first
100msec of the stimulus presentation, which was transduced through the
platform’s load cell and expressed in arbitrary units (AU).  

\subsubsection{Somatosensation}
Mouse somatosensation was tested by placing mice in a startle
chamber (Med Associates) connected to Startle Pro Software. Mice were placed
atop a multi-bar cradle within a ventilated plexiglass restraint cylinder, which
allows for horizontal movement within the chamber, but not upright rearing.
After two minutes of acclimation, each mouse was exposed to 10 different
scrambled shock intensities, ranging from 0 to 0.6mA with randomized inter-trial
intervals of 20-90 seconds. Each animal’s startle reactivity during the first
100 msec of the shock was transduced through the platform’s load cell and
recorded as area under the curve (AUC) in arbitrary units (AU). 

% QUANTIFICATION AND STATISTICAL ANALYSIS
\section{Quantification and statisical analysis}

Experimental conditions, number of replicates, and statistical tests used are
stated in each figure legend. Each experiment was replicated at least three
times (or on at least 3 separate animals) to assure rigor and reproducibility.
Both male and female age-matched mice were used for all experiments, with data
pooled from both sexes. Data compilation and statistical analyses for all
non-proteomic data were performed using GraphPad Prism (version 8, GraphPad
Software, CA), using a significance level of alpha=0.05. Prism provides exact p
values unless p<0.0001. All data are reported as mean ± SEM. Each data set was
tested for normal distribution using a D'Agostino-Person normality test to
determine whether parametric (unpaired Student's t-test, one-way ANOVA, two-way
ANOVA) or non-parametric (Mann-Whitney, Kruskal-Wallis) tests should be used.
Parametric assumptions were confirmed with the Shapiro-Wilk test (normality) and
Levine's test (error variance homogeneity) for ANOVA with repeated measures
testing. The analysis of iBioID and TMT proteomics data are described below. All
proteomic data and analysis scripts are available online (see Resource
Availability).

\subsection{Imaris 3D reconstruction}

For EEA1+ and CathepsinD+ puncta analyses, coverslips were imaged on a Zeiss LSM
710 confocal microscope. Images were sampled at a resolution of 1024 x 1024
pixels with a dwell time of 0.45µsec using a 63x/1.4 oil immersion objective, a
2.0 times digital zoom, and a z-step size of 0.37 µm. Images were saved as
“.lsm” formatted files, and quantification was performed on a POGO Velocity
workstation in the Duke Light Microscopy Core Facility using Imaris 9.2.0
software (Bitplane, South Windsor, CT). For analyses, we first used the
“surface” tool to make a solid fill surface of the MAP2-stained neuronal soma
and dendrites, with the background subtraction option enabled. We selected a
threshold that demarcated the neuron structure accurately while excluding
background. For EEA1 puncta analyses, a 600 x 800 µm selection box was placed
around the soma in each image and surfaces were created for EEA1 puncta within
the selection box. Similarly, for CathepsinD puncta analyses, a 600 x 600 µm
selection box was placed around the soma(s) in each image for surface creation.

The same threshold settings were used across all images, and individual surface
data from each soma were exported for aggregate analyses. The experimenter was
blinded to sample conditions for both image acquisition and analysis.

\subsection{Cleaved Caspase-3 Image Analysis}

Z-stack images were acquired on a Zeiss 710 LSM confocal microscope. Images were
sampled at a resolution of 1024 x 1024 pixels with a dwell time of 1.58µsec,
using a 63x/1.4 oil immersion objective (for cortex, striatum, and hippocampus)
or 20x/0.8 dry objective (cerebellum), a 1.0 times digital zoom, and a z-step
size of 0.67 µm. Images were saved as “.lsm” formatted files, and then converted
into maximum intensity projections (MIP) using Zen 2.3 SP1 software.

Quantification of CC3 colocalization with DAPI was performed on the MIPs using
the Particle Analyzer function in FIJI ImageJ software. The experimenter was
blind to sample conditions for both image acquisition and analysis.

\subsection{iBioID Quantitative Analysis}

Following UPLC-MS/MS analyses, data was imported into Proteome Discoverer 2.2
(Thermo Scientific Inc.), and aligned based on the accurate mass and retention
time of detected ions (“features”) using Minora Feature Detector algorithm in
Proteome Discoverer. Relative peptide abundance was calculated based on
area-under-the-curve (AUC) of the selected ion chromatograms of the aligned
features across all runs. The MS/MS data was searched against the SwissProt Mus
musculus database (downloaded in April 2018) with additional proteins, including
yeast ADH1, bovine serum albumin, as well as an equal number of
reversed-sequence “decoys” for false discovery rate (FDR) determination. Mascot
Distiller and Mascot Server (v 2.5, Matrix Sciences) were utilized to produce
fragment ion spectra and to perform the database searches. Database search
parameters included fixed modification on Cys (carbamidomethyl), variable
modifications on Meth (oxidation) and Asn and Gln (deamidation), and were
searched at 5 ppm precursor and 0.02 Da product mass accuracy with full trypsin
enzymatic rules. Peptide Validator and Protein FDR Validator nodes in Proteome
Discoverer were used to annotate the data at a maximum 1\% protein FDR.

Protein intensities were exported from Proteome Discoverer and processed using
custom R scripts. Carboxylases and keratins, as well as 315 mitochondrial
proteins(Calvo et al., 2016), were removed from the identified proteins as known
contaminants.

Next, we performed sample loading normalization to account for
technical variation between the 9 individual MS runs. This is done by
multiplying intensities from each MS run by a scaling factor, such that the
average of all total run intensities are equal. As QC samples were created by
pooling equivalent aliquots of peptides from each biological replicate, the
average of all biological replicates should be equal to the average of all
technical SPQC replicates. We performed sample pool normalization to SPQC
samples to standardize protein measurements across all samples and correct for
batch effects between MS analyses. Sample pool normalization adjusts the
protein-wise mean of all biological replicates to be equal to the mean of all
SPQC replicates. Finally, proteins that were identified by a single peptide,
and/or identified in less than 50\% of samples were removed. Any remaining
missing values were inferred to be missing not at random due to the left shifted
distribution of proteins with missing values and imputed using the k-nearest
neighbors algorithm using the impute.knn function in the R package impute
(impute::impute.knn). Normalized protein data was analyzed using edgeR, an R
package for the analysis of differential expression/abundance that models count
data using a binomial distribution methodology. Differential enrichment of
proteins in the WASH1-BioID2 pull-down relative to the solubleBioID2 control
pull-down were evaluated with an exact test as implemented by the
edgeR::exactTest function. To consider a protein enriched in the WASH
interactome, we required that a protein exhibit a fold change greater than 3
over the negative control with an exact test Benjamini Hochberg adjusted p-value
(FDR) less than 0.1. With these criteria, 174 proteins were identified as WASH1
interactome proteins. Raw peptide and final normalized protein data as well as
the statistical results can be found in Table S1.

Proteins that function together often interact directly. We compiled
experimentally-determined protein-protein interactions (PPIs) among the WASH1
interactome from the HitPredict database(López et al., 2015) using a custom R
package, getPPIs, (available online at twesleyb/getPPIs). We report PPIs among
the WASH1 interactome in Table S1.

Bioinformatic GO analysis was conducted by manual annotation of identified
proteins and confirmed with Metascape analysis(Zhou et al., 2019) of
WASH1-BioID2 enriched proteins using the 2,311 proteins identified in the mass
spec analysis as background.

Raw peptide intensities were exported from Proteome Discover for downstream
analysis and processing in R. Following database searching, protein scoring
using the Protein FDR Validator algorithm, and removal of contaminant species,
the dataset retained 86,551 peptides corresponding to the identification of
7,488 unique proteins. These data, as well as statistical results can be found
in Table S2.

\subsection{TMT Proteomics Quantitative Analysis}

Peptide level data from the
spatial proteomics analysis of SWIP\textsuperscript{P1019R} MUT and MUT brain
were exported from Proteome Discoverer (version 2.4) and analyzed using custom R
and Python scripts.

Peptides from contaminant and non-mouse proteins were
removed. First, we performed sample loading normalization, normalizing the total
ion intensity for each TMT channel within an experiment to be equal. Sample
loading normalization corrects for small differences in the amount of sample
analyzed and labeling reaction efficiency differences between individual TMT
channels within an experiment.

We found that in each TMT experiment there were a small number of missing values
(mean percent missing = 1.6 +/- 0.17\%). Missing values were inferred to be
missing at random based on the overlapping distributions of peptides with
missing values and peptides without missing values. We imputed these missing
values using the k-nearest neighbor algorithm (impute::impute.knn). Missing
values for SPQC samples were not imputed. Peptides with any missing SPQC data
were removed.

Following sample loading normalization, SPQC replicates within each experiment
should yield identical measurements. As peptides with irreproducible QC
measurements are unlikely to be quantitatively robust, and their inclusion may
bias downstream processing (see IRS normalization below), we sought to remove
them. To assess intra-batch variability, we utilized the method described by
Ping et al., 2019(Ping et al., 2018). Briefly, peptides were binned into 5
groups based on the average intensity of the two SPQC replicates. For each pair
of SPQC measurements, the log ratio of SPQC intensities was calculated. To
identify outlier QC peptides, we plotted the distribution of these log ratios
for each bin. Peptides with ratios that were more than four standard deviations
away from the mean of its intensity bin were considered outliers and removed
(Total number of SPQC outlier peptides removed = 474).

Proteins were summarized as the sum of all unique peptide intensities
corresponding to a unique UniProtKB Accession identifier, and sample loading
normalization was performed across all three experiments to account for
inter-experimental technical variability. In a TMT experiment, the peptides
selected for MS2 fragmentation for any given protein is partially random,
especially at lower signal-to-noise peptides. This stochasticity means that
proteins are typically quantified by different peptides in each experiment.
Thus, although SPQC samples should yield identical protein measurements in each
of the three experiments (as it is the same sample analyzed in each experiment),
the observed protein measurements exhibit variability due to their
quantification by different peptides. To account for this protein-level bias, we
utilized the internal reference scaling (IRS) approach described by Plubell et
al., 2017(Plubell et al., 2017). IRS normalization scales the protein-wise
geometric average of all SPQC measurements across all experiments to be equal,
and simultaneously adjusts biological replicates. In brief, each protein is
multiplied by a scaling factor which adjusts its intra-experimental SPQC values
to be equal to the geometric mean of all SPQC values for the three experiments.
This normalization step effectively standardizes protein measurements between
different mass spectrometry experiments.

The final normalization step was to perform sample pool normalization using SPQC
samples as a reference. This normalization step, sometimes referred to as global
internal standard normalization, accounts for batch effects between experiments,
and reflects the fact that after technical normalization, the mean of biological
replicates should be equal to the mean of SPQC replicates.

Before assessing protein differential abundance, we removed irreproducible
proteins. This included proteins that were quantified in less than 50\% of all
samples, proteins that were identified by a single peptide, and proteins that
had missing SPQC values.

Across all 42 biological replicates, we observed that
a small number of proteins had potential outlier measurements that were either
several orders of magnitude greater or less than the mean of its replicates. In
order to identify and remove these proteins, we assessed the reproducibility of
protein measurements within a fraction in the same manner used to identify and
filter SPQC outlier peptides. A small number of proteins were identified as
outliers if the average log ratio of their 3 technical replicates was more than
4 standard deviations away from the mean of its intensity bin (n=349). In total,
we retained 5,897 of the original 7,488 proteins in the final dataset.

Differential protein abundance was assessed using the final normalized protein
data for intrafraction comparisons between WT and MUT groups using a general
linear model as implemented by the edgeR::glmQLFit and edgeR::glmQLFTest
functions(MD et al., 2009). Although this approach was originally devised for
analysis of single-cell RNA-sequencing data, this approach is also appropriate
for proteomics count data which is over-dispersed, negative binomially
distributed, and often only includes a small number of replicates (for an
example of edgeR's application to proteomics see Plubell et al., 2017(Plubell et
al., 2017))(McCarthy et al., 2012; MD et al., 2009). For intrafraction
comparisons, P-values were corrected using the Benjamini Hochberg procedure
within edgeR. An FDR threshold of 0.1 was set for significance for intrafraction
comparisons.

We utilized edgeR's flexible GLM framework to test the hypothesis that the
abundance of proteins in the WT group was significantly different from that in
the MUT group irrespective of fraction differences (Table S2). For WT vs. MUT
contrasts, we considered proteins with an FDR < 0.05 significant (n=687). For
plotting, we adjusted normalized protein abundances for fraction differences by
fitting the data with an additive linear model with fraction as a blocking
factor, as implemented by the removeBatchEffect algorithm from the R limma
package(Ritchie et al., 2015).

To construct a protein covariation graph, we assessed the pairwise covariation
(correlation) between all 5,897 proteins using the biweight midcorrelation
(WGCNA::bicor) statistic(Seyfried et al., 2017), a robust alternative to
Pearson's correlation. The resulting complete, signed, weighted, and symmetric
adjacency matrix was then re-weighted using the 'Network Enhancement' approach.
Network enhancement removes noise from the graph, and facilitates downstream
community detection(Wang et al., 2018).

The enhanced adjacency matrix was clustered using the Leiden algorithm(Traag et
al., 2019), a recent extension and improvement of the well-known Louvain
algorithm(Mucha et al., 2010). The Leiden algorithm functions to optimize the
partition of a graph into modules by maximizing a quality statistic. We utilized
the 'Surprise' quality statistic(Traag et al., 2015) to identify optimal
partitions of the protein covariation graph. To facilitate biological inferences
drawn from the network's organization, we recursively split 27 modules that
contained more than 100 nodes and removed modules that were smaller than 5
proteins. Initial clustering of the network resulted in the identification of
324 modules.

To reduce the likelihood of identifying false positive modules, we enforced
module quality using a permutation procedure
(NetRep::modulePreservation)(Ritchie et al., 2016) and removed modules with any
insignificant permutation statistics (Bonferroni P-Adjust > 0.05). The following
statistics were used to enforce module quality: 'avg.weight' (average edge
weight), 'avg.cor' (average bicor correlation R2), and 'avg.contrib' (quantifies
how similar an individual protein's abundance profile is to the summary of its
module). Proteins which were assigned to modules with insignificant module
quality statistics were not considered clustered as the observed quality of
their module does not differ from random.

After filtering, approximately 85\%
of all proteins were assigned a cluster. The median percent variance explained
by the first principle component of a module (a measure of module cohesiveness)
was high (59.8\%). After removal of low-quality modules, the analysis retained
255 distinct modules of proteins that strongly covaried together (Table S3).

To evaluate modules that were changing between WT and MUT genotypes, we extended
the GLM framework to test for protein differential abundance. Modules were
summarized as the sum of their proteins and fit with a GLM, with fraction as a
blocking factor. In this statistical design, we were interested in the average
effect of genotype on all proteins in a module. For plotting, module abundance
was adjusted for fraction differences using the removeBatchEffect function
(package: limma).

We utilized the Bonferroni method to adjust P-values for 255 module level
comparisons and considered modules with an adjusted P-value less than 0.05 were
considered significant (n=37).

\subsection{Module Gene Set Enrichment Analysis}
Modules were analyzed for enrichment of the WASH interactome (this paper), Retriever complex (McNally et al., 2017), CORUM
protein complexes (Giurgiu et al., 2019), and subcellular predictions generated
by Geladaki et al.(Geladaki et al., 2019) using the hypergeometric test with
Bonferroni P-value correction for multiple comparisons.

The union of all
clustered and pathway proteins was used as background for the hypergeometric
test. In addition to analysis of these general cellular pathways, we analyzed
modules for enrichment of neuron-specific subcellular compartments—this included
the presynapse (Takamori et al., 2006), excitatory postsynapse (Uezu et al.,
2016), and inhibitory postsynapse (Uezu et al., 2016). These gene lists are
available online at https://github.com/twesleyb/geneLists.

\subsection{Network Visualization}
Network graphs were visualized in Cytoscape (Version 3.7.2). We
used the Perfuse Force Directed Layout (weight = edge weight). In this layout,
strongly connected nodes tend to be positioned closer together. In some
instances, node location was manually adjusted to visualize the module more
compactly. Node size was set to be proportional to the weighted degree
centrality of a node in its module subgraph. Node size thus reflects node
importance in the module. Visualizing co-expression or co-variation networks is
challenging because every node is connected to every other node (the graph is
complete). To aid visualization of module topology, we removed weak edges from
the graphs. A threshold for each module was set to remove the maximal number of
edges before the module subgraph split into multiple components. This strategy
enables visualization of the strongest paths in a network.
