% title: response.tex
% description: Response to eLife Reviewers
% author: twab

% USAGE: 
% to compile this document:
%    R  >>> knitr::knit("response.Rnw")
%    sh >>> pdflatex response.tex


%% R --------------------------------------------------------------------------
<<renv, echo=FALSE, cache=FALSE, results='hide'>>=

# prepare the R environment
root <- "~/projects/SwipProteomics"
renv::load(root,quiet=TRUE)
devtools::load_all(quiet=TRUE)

suppressPackageStartupMessages({
  library(knitr)
  library(data.table)
})

# global knitr options
defaults <- list(tidy=FALSE, message=FALSE, warning=FALSE, 
	     fig.width=6, fig.height=4.5, results='hide')
do.call(knitr::opts_chunk$set, defaults)

# set code highlighting theme
knitr::knit_theme$set("bclear") # good options: acid, bclear, earendel, bright

# source external R code
knitr::read_chunk('./R/supplement.R')

@


%% latex document setup -------------------------------------------------------

\documentclass[11pt]{elife}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsthm}
\usepackage{ragged2e}
\usepackage{caption}
\usepackage{fancyhdr}
\usepackage{graphicx}
\usepackage{titlesec}
\usepackage{blkarray}
\usepackage{csquotes}

\graphicspath{ {./figs/} }


\title{Supplementary Statistical Methods\\
\small{Genetic Disruption of WASHC4 Drives Endo-lysosomal Dysfunction and \\
Cognitive-Movement Impairments in Mice and Humans}}

\author[1\authfn{0}]{Jamie Courtland}
\author[1\authfn{0}]{Tyler W. A. Bradshaw}
\author[2]{Greg Waitt}
\author[2,3]{Erik J. Soderblom}
\author[2]{Tricia Ho}
\author[4]{Anna Rajab}
\author[5]{Ricardo Vancini}
\author[2\authfn{1}]{Il Hwan Kim}
\author[6]{Ting Huang}
\author[6]{Olga Vitek}
\author[3]{Scott H. Soderling}

\affil[1]{Department of Neurobiology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[2]{Proteomics and Metabolomics Shared Resource, 
Duke University School of Medicine, Durham, NC 27710, USA}
\affil[3]{Department of Cell Biology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[4]{Burjeel Hospital, VPS Healthcare, Muscat, Oman}
\affil[5]{Department of Pathology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[6]{Khoury College of Computer Sciences, Northeastern University,
Boston, MA 02115, USA}

\contrib[\authfn{0}]{These authors contributed equally to this work.}
\presentadd[\authfn{1}]{Department of Anatomy and Neurobiology, 
University of Tennessee Health Science Center, Memphis, TN 38163, USA}

\corr{jamie.courtland@duke.edu}{JC}
\corr{tyler.w.bradshaw@duke.edu}{TWAB}
\corr{greg.waitt@duke.edu}{GW}
\corr{erik.soderblom@duke.edu}{EJB}
\corr{tricia.ho@duke.edu}{TH}
\corr{drannarajab@gmail.com}{DR}
\corr{ricardo.vancini@duke.edu}{RV}
\corr{ikim9@uthsc.edu}{IK}
\corr{huang.tin@northeastern.edu}{TH}
\corr{o.vitek@northeastern.edu}{OV}
\corr{scott.soderling@duke.edu}{SHS}


\setlength{\abovedisplayskip}{3pt}
\setlength{\belowdisplayskip}{3pt}


%% main -----------------------------------------------------------------------

\begin{document}

\maketitle

\renewcommand{\abstractname}{Summary}
\begin{abstract}

Here we address concerns about the statistical validity of our previous approach
to assess differential protein abundance in the \textbf{WASH-iBioID} and
\textbf{SWIP-TMT} proteomics datasets. Our previous approach depended
upon the R package \texttt{edgeR}. We used \texttt{edgeR} to perform
both protein- and module-level inference---assessing differential
abundance of individual proteins as well as protein groups in
SWIP\textsuperscript{P1019R} mouse brain. \texttt{edgeR} utilizes a
negative binomial (NB) statistical framework originally developed for
analysis of RNA-Seq read count data. Previously, we failed to fully
consider the validity of \texttt{edgeR's} NB assumption for proteomics
data. We evaluate the goodness-of-fit of the negative binomial model for
our TMT dataset and find evidence of a lack-of-fit.  Thus, we	revise
our statistical approach and reanalyze our data, making use of
\cite{Huang2020}'s recently published R package \texttt{MSstatsTMT}.
\texttt{MSstatsTMT} uses a flexible linear mixed-model (LMM) statistical
framework which we extend to re-evaluate both protein- and module-level
statistical comparisions in our SWIP-TMT spatial proteomics dataset.

\end{abstract}

\newpage


\section{Goodness-of-fit of the NB Model for TMT MS}

Our previous method can be summarized as the \textit{Sum + IRS} approach
\citep{Huang2020}.  Following protein summarization 
and internal reference scaling (IRS) normalization \citep{Plubell2017},  we
applied \texttt{edgeR} \citep{McCarthy2012} to assess differential abundance of
individual proteins and protein-groups.  The use of \texttt{edgeR} for
protein-level comparisons was based on work by \cite{Plubell2017} who describe
IRS normalization and the use of \texttt{edgeR} for statistical testing in TMT
MS experiments \citep{Plubell2017}.  We failed however, to consider the overall
adequecy of \texttt{edgeR's} NB GLM model for our TMT proteomics data.

Statistical inference in \texttt{edgeR} is performed for each gene or protein
using a negative binomial, generalized linear model framework.  The data are
assumed to be adequately described by a NB distribution parameterized by a
dispersion parameter, $\phi$. The dispersion parameter $\phi$ accounts for
mean-variance relationships in proteomics and transcriptomics data.  As signal
intensity in protein MS is fundamentally related to the number of ions generated
from an ionized, fragmented protein, we incorrectly inferred that TMT mass
spectrometry data can be modeled as NB count data. Based on this assumption, we
justified our use of \texttt{edgeR}.  

To evaluate the overall adequacy of the negative binomial model for TMT
proteomics data, we plot the residual protein deviance statistics of all
proteins fit with \texttt{edgeR's} NB GLM against their theoretical normal
quantiles in a quantile-quantile (QQ) plot (FIG:gof).  The QQ plot addresses the
question of how similar the observed data are to the theoretical distribution
given by the NB model.  A linear relationship between the observed and
theoretical values is a goodness-of-fit indicator.  Deviation from this linear
trend is evidence of a lack-of-fit.

Following protein summarization and normalization with \texttt{MSstatsTMT}, the
SWIP-TMT data were fit with a NB GLM using \texttt{edgeR::glmFit}. \FIG{gof}
illustrates the divergence of the observed and theoretical quantiles for our
SWIP-TMT dataset fit with \texttt{edgeR's} NB GLM. Given our experimental
design, \texttt{MSstatsTMT} fits an appropriate linear-mixed model to the data.
The quantile-quantile plot in \FIG{gof} indicates that the data are well
described by \texttt{MSstatsTMT's} LMM, which does not depend upon the negative
binomial assumption.


\section{Protein-wise Linear Mixed-Models}

\cite{Huang2020} created \texttt{MSstatsTMT}, an R package for data
normalization and hypothesis testing in multiplex TMT proteomics experiments.
\texttt{MSstatsTMT} performs statistical inference in two steps.
First, each protein in the dataset is fit with a LMM
expressing the major sources of variation in the experimental design.
Second, given the fitted model, a model-based comparison is made between pairs
of treatment conditions. Using LMMs we can untangle the variance
attributable to the biological effect we are interested in from the experimental
and biological covariates which mask this response.

\cite{Huang2020} outline a common vocabulary for describing TMT MS experimental
design. An experiment consists of \texttt{m = 1} ...  \texttt{M}\ concatenations
of isobarically labeled samples or \texttt{Mixtures}.  This mixture is then
analyzed by the mass spectrometer in a single MS \texttt{Run}.  This mixture is
often fractionated into multiple liquid chromotography \texttt{Fractions} to
decrease sample complexity, and thereby increase the depth of proteome coverage.
Within a mixture, each of the unique TMT channels is dedicated to the analysis
of \texttt{c = 1} ...  \texttt{C}\ individual biological or treatment
\texttt{Conditions}.  There may then be \texttt{b = 1} ...  \texttt{B}\
biological replicates or \texttt{Subjects}.  Finally, a single TMT mixture may
be repeatedly analyzed in \texttt{t = 1} ...  \texttt{T}\ technical replicate
mass spectrometry runs.

Equation \ref{eq:full} is a LMM describing protein abundance as a function of
the major sources of variation  in a general TMT experiment composed of
\texttt{M} mixtures, \texttt{T} technical replicates of mixture, \texttt{C}
conditions, and \texttt{B} biological subjects.
\begin{equation} % eq:full
  \label{eq:full} 
	Y_{mcbt} = \mu + Condition_c + Mixture_m + TechRep(Mixture)_{m(t)} + 
	Subject_{mcb} + \epsilon_{mcbt}\\
\end{equation}

\begin{equation}
  \begin{gathered}
    \label{eq:constraints}
	\sum_{c=1}^{C} Condition_c = 0 \\
	Mixture_m \stackrel{iid}{\sim} N(0,\sigma^2_M) \\
	TechRep(Mixture)_{t(m)} \stackrel{iid}{\sim} N(0,\sigma^2_T) \\
	Subject_{mcb} \stackrel{iid}{\sim} N(0,\sigma^2_S) \\
	\epsilon{mtcb} \stackrel{iid}{\sim} N(0,\sigma^2) \\
  \end{gathered}
\end{equation}

The model's constraints \ref{eq:constraints} distinguish fixed- and mixed-effect
components of variation in the response, $Y_{mcbt}$. \texttt{Mixture} is a
mixed-effect and represents variation between different TMT mixtures. By
definition mixed-effects are assumed to be normally and independently
distributed (\texttt{iid}).  The term \texttt{TechRep(Mixture)} represents
random variation between replicates of a single MS \texttt{Run}.
\texttt{Subject} corresponds to each unique biological replicate and represents
biological variation among the levels of the fixed-effect term \texttt{Condition}.
The term $\epsilon_{mtcb}$ is a mixed-effect representing both biological and
technical variation, quantifying any remaining error. If a component of the
model is not estimable, then it is removed.  For example, if there is no
technical replication of mixture \texttt{(T=0)}, then the model is reduced to
equation \ref{eq:reduced}.
\begin{equation} % NOTE: dont put blank lines above equations!
	\label{eq:reduced} % equation -- reduced
	Y_{mcbt} = \mu + Condition_c + Mixture_m + Subject_b + \epsilon_{mcb}
\end{equation}


\section{SWIP-TMT Spatial Proteomics}

We analyzed the brains of mice with the SWIP\textsuperscript{P1019R} mutation by
subcellular fractionation and TMT MS profiling.  We aimed to reveal how this
pathogenic mutation may perturb the organization of the subcellular proteome.
We adapted the subcellular fractionation method of \cite{Geladaki2019} to
prepare seven subcellular fractions from the brains of control and
SWIP\textsuperscript{P1019R} mutant mice.  Our experimental design is summarized
in \FIG{design}.  

Each 16-plex TMT \texttt{Mixture} was composed of fourteen
biological fractions or \texttt{BioFractions} obtained from subcellular
fractionation of a control and SWIP\textsuperscript{P1019R} mutant mouse brain.
We refer to these subcellular preparations as a \texttt{BioFractions} to
distinguish them from an MS \texttt{Fraction}.  The term \texttt{Condition} of
equation \ref{eq:reduced} represents these fourteen combinations of
\texttt{Genotype} and \texttt{BioFraction}.  

In our design, \texttt{Mixture} is confounded with \texttt{Subject}. We analyzed
all seven \texttt{BioFractions} from a single control and mutant mouse in the
same \texttt{Mixture}.  We choose to model the effect of \texttt{Mixture} and
not \texttt{Subject} based on the assumption that the experimental batch effect
represented by the term \texttt{Mixture} is greater than the error inherent in
the repeated measures of each \texttt{Subject}.  We omit the un-estimable terms
\texttt{TechRep(Mixture)} and \texttt{Subject} from equation (\ref{eq:full}).
The reduced linear mixed-model describing our experimental design is given by
equation \ref{eq:fx0}.
\begin{equation}
	\label{eq:fx0}
	Y_{mcbt} = \mu + Condition_c + Mixture_m + \epsilon_{mcb}
\end{equation}


\section{Statistical Inference with MSstatsTMT}

\texttt{MSstatsTMT} performs protein-wise comparisons between pairs of
\texttt{Conditions} by comparing the estimates obtained from the LMM fit by
restricted maximum likelihood \citep{Bates2015}. We are interested in testing
the hypothesis:
\begin{equation}
	\label{eq:null} % equation -- null
	H0 : l^T * \beta = 0 
\end{equation}

Where $l^T$ is a vector of $\sum=1$ specifying the positive and negative
coefficients of a contrast. $\beta$ is the model-based estimates of
\texttt{Condition}.  The null hypothesis (\ref{eq:null}) is that the fold
change, $\l^T * \beta$, is 0.  A test statistic for such a two-way contrasts is
given by \cite{Kuznetsova2017}:
\begin{equation} 
	\label{eq:tstatistic} % equation -- tstatistic
	t = \frac{l^T \hat{\beta}}{\sqrt{l \sigma^2 \hat{V} l^T}}
\end{equation}

We obtain the model's estimates, $\hat{\beta}$, error,$\sigma^2$, and
variance-covariance matrix,$\hat{V}$, from the fit LMM.  Given a contrast, $l^T$,
the numerator of equation (\ref{eq:tstatistic}) is the fold change of a
comparison.  The product of $\sigma^2$ and $\hat{V}$ is the scaled
variance-covariance matrix describing error estimates of the model's fixed- and
mixed-effect parameters.  Together the denominator represents the standard error
of the comparison. The degrees of freedom for the contrast are derived using the
Satterthwaite moment of approximation method \citep{Kuznetsova2017}.  Finally, a
p-value is calculated given the t-statistic and degrees of freedom.  P-values
for the protein-wise tests are adjusted using the Benjamini-Hochberg FDR method
\citep{Huang2020}.

We used \texttt{MSstatsTMT} to assess two types of contrasts.
\texttt{Intra-BioFraction} comparisons are the seven pairwise comparisons of
control and mutant protein abundance for each \texttt{BioFraction}.  We also
assessed the overall \texttt{Mutant-Control} comparison.  Each of these
contrasts is represented by a vector, $l^T$, which specifies a contrast between
coefficients of \texttt{Condition} in the LMM (\ref{eq:fx0}).  \FIG{contrasts}
illustrates the two types of protein-level statistical comparisons we implement
with \texttt{MSstatsTMT}.


\section{Module-level Inference with Mixed-Models}

The strength of linear mixed-models lies in their flexibility. In a mixed-model
the response variable is taken to be a function of both fixed- and
random-effects.  If the set of possible levels of a covariate is fixed and
reproducible, then the factor is modeled as a fixed-effect parameter.  In
contrast, if the levels of an observation reflect a sampling of the set of all
possible levels, then the covariate is modeled as a random-effect.  Random or
mixed-effects represent categorical variables that reflect experimental or
observational units within the dataset.  As such, mixed-effect parameters
account for the variation occurring among lower levels of an upper level unit in
the data \citep{Bates2015}.  

We wish to extend the LMM framework developed by \texttt{MSstatsTMT} to perform
inference at the level of protein groups. Given a map partitioning the proteome
into modules of covarying proteins, we wish to assess the module-level
difference between control and SWIP\textsuperscript{P1019R} conditions. We fit
the data for each module in the dataset with a LMM. We represent the proteins
within each module as the mixed-effect term \texttt{Protein}, capturing
variation among a module's constintuent proteins.
\begin{equation} 
  \begin{gathered}\label{eq:fx1} % equation -- fx1
	Y_{mcbt} = \mu + Condition_c + Mixture_m + Protein_p + \epsilon_{mcb}\\
	Protein_p \stackrel{iid}{\sim} N(0,\sigma^2_P) \\
  \end{gathered}
\end{equation}

The term \texttt{Protein} in equation \ref{eq:fx1} quantifies the variance
$\sigma_P$ attributable to heterogenity among a modules proteins.


\section{LMM Goodness-of-fit}

It is useful to consider the goodness-of-fit of our models. A straight forward
measure of a LMM's quality is the Nakagawa coefficient of 
determination \citep{Nakagawa2012}. \cite{Nakagawa2012}'s conditional $R^2_c$ is 
interpreted as the total variance explained by a LMM ($R^2_{total}$).
The marginal $R^2_m$ is interpreted as the variance explained by the LMM's 
fixed-effects ($R^2_{fixed}$). We implement \cite{Nakagawa2012}'s coeffficient of 
determination using the \texttt{r.squaredGLMM} function taken from the 
\texttt{MuMin} package \citep{WangMerkle2018}. 

In addition to considering the total variance explained by a module, it is
helpful to consider the variance explained by each of its factors.
The R package \texttt{variancePartition} enables us to
calculate the percent variance explained by a LMM's parameters
cite(variancePartition). 


\section{Spatial Proteomics Network Construction}

Using our SWIP-TMT dataset, we aim to identify modules or groups of
proteins that covary together across subcellular space. Prior to building the
co-variation network, other sources of variation should be removed. Although
\texttt{MSstatsTMT} handles the batch effect inherent in experiments with
multiple TMT mixtures, it is necessary to remove this effect prior to building
the network. We removed the effect of \texttt{Mixture} using
\texttt{limma::RemoveBatchEffect}. These adjusted data are used for network
construction and plotting, but not statistical modeling.

Prior to network construction, we removed protein models with poor fit 
($R^2_{total}<0.7$; n=791 proteins). Removing this noisey proteins facilitation
module identification and improves overall module quality.

The final network was constructed using data from both Control and Mutant 
samples after adjusting for batch (Mixture). The final dataset included 
42 samples and 6,119 proteins. The protein covariation network was build by
calculating the Pearson correlation for all pairwise comparisons of proteins.

We performed network enhancement to remove biological noise from the network
prior to clustering. This step is essential in large, and dense networks for
module detection. Network enhancement reweights the
network's edges and has the overall effect of making the network sparse.
Conceptually this step is related to the soft-thresholding approach taken by
WGCNA or WPCNA analysis workflows (REFS), but has the benefit of not assuming
that the network has an overall scale-free topology.  Without reweighting or
enhancing the network, most extant clustering algorithms fail to detect
communities in the dataset.  Network enhancement has the effect of making the
network sparse and facilitates the identification of network structure.\\


\section{Community Detection with Leidenalg}

To reveal the structure of our spatial proteomics network we used the recently
published Leiden algorithm \citep{Traag2019}.

%------------------------------------------------------------------------------
\section{Implementation}

In order to understand and extend the function of \texttt{MSstatsTMT}, we
extracted \texttt{MSstatsTMT's} core model-fitting and statistical testing
steps.  At the core of the model fitting-step is the R package \texttt{lme4}
which implements mixed-effects models with its function
\texttt{lme4::lmer}\citep{Bates2015}. The package \texttt{lmerTest} extends
\texttt{lme4's} functionality and enables the computation of Sattertwaite
degrees of freedom \citep{Kuznetsova2017}. 

\paragraph{Fit WASHC4}

As a means of example, we demonstrate the model fitting statistical testing
steps for both and protein- and module-level statistical comparisons.
First, we fit the LMM in (\ref{eq:fx0}) the normalized protein level data from
\texttt{MSstatsTMT} for WASHC4.
<<fit0, results='markup'>>=
@

\paragraph{'MutantF7-ControlF7'}

We assess the contrast between \texttt{BioFraction} seven (F7) mutant and
control conditions.
<<contrast7, results='markup'>>=
@

\paragraph{'Mutant-Control'}

Provided the correct contrast, we easily evaluate the overall difference
between mutant and control mice.
<<contrast8, results='markup'>>=
@

\paragraph{Fit WASH Complex}
Next we fit a LMM to the five WASH complex proteins.
<<fit1, results='markup'>>=
@

We evaluate the goodness of fit of our module-level model.
<<nakagawa, results='markup'>>=
@

\paragraph{varianceParition}
We compute the variance explained using \texttt{variancePartition}.
<<variancePartition>>=
@

%------------------------------------------------------------------------------
\newpage


%\section{References}

\bibliography{bibliography}

\newpage

\section{Supplemental Tables}

\begin{figure}
<<tab0, fig.align='center',echo=FALSE, results="markup">>=
@
	\caption{This is a caption.}
	\label{fig:tab0}
\end{figure}

\newpage


\section{Supplemental Figures}

\begin{itemize}
	\item gof
	\item design
	\item contrasts
	\item ...
\end{itemize}

\newpage

% figures
\include{figures}


\end{document}
