% title: response.tex
% description: Response to eLife Reviewers
% author: twab

% USAGE: 
% to compile this document:
%    R  >>> knitr::knit("response.Rnw")
%    sh >>> pdflatex response.tex


<<renv, echo=FALSE, cache=FALSE, results='hide'>>=

# prepare the R environment
root <- "~/projects/SwipProteomics"
renv::load(root,quiet=TRUE)
devtools::load_all(quiet=TRUE)

suppressPackageStartupMessages({
  library(knitr)
  library(dplyr)
  library(data.table)
})

# global knitr options
opts <- list(tidy=FALSE,message=FALSE,warning=FALSE,fig.width=4,fig.height=4.5)
do.call(opts_chunk$set,opts)

@


% latex document setup

\documentclass[11pt]{elife}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsthm}
\usepackage{ragged2e}
\usepackage{caption}
\usepackage{fancyhdr}
\usepackage{graphicx}
\usepackage{titlesec}
\usepackage{blkarray}
\usepackage{csquotes}

% location of figures
\graphicspath{ {./figs/} }


\title{Supplementary Methods\\
\small{Genetic Disruption of WASHC4 Drives Endo-lysosomal Dysfunction and \\
Cognitive-Movement Impairments in Mice and Humans}}


% Authors
\author[1\authfn{0}]{Jamie Courtland}
\author[1\authfn{0}]{Tyler W. A. Bradshaw}
\author[2]{Greg Waitt}
\author[2,3]{Erik J. Soderblom}
\author[2]{Tricia Ho}
\author[4]{Anna Rajab}
\author[5]{Ricardo Vancini}
\author[2\authfn{1}]{Il Hwan Kim}
\author[6]{Ting Huang}
\author[6]{Olga Vitek}
\author[3]{Scott H. Soderling}

\affil[1]{Department of Neurobiology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[2]{Proteomics and Metabolomics Shared Resource, 
Duke University School of Medicine, Durham, NC 27710, USA}
\affil[3]{Department of Cell Biology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[4]{Burjeel Hospital, VPS Healthcare, Muscat, Oman}
\affil[5]{Department of Pathology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[6]{Khoury College of Computer Sciences, Northeaster University,
Boston, MA 02115, USA}

\contrib[\authfn{0}]{These authors contributed equally to this work.}
\presentadd[\authfn{1}]{Department of Anatomy and Neurobiology, 
University of Tennessee Health Science Center, Memphis, TN 38163, USA}

\corr{jlc123@duke.edu}{JC}
\corr{tyler.w.bradshaw@duke.edu}{TWAB}
\corr{greg.waitt@duke.edu}{GW}
\corr{erik.soderblom@duke.edu}{EJB}
\corr{tricia.ho@duke.edu}{TH}
\corr{drannarajab@gmail.com}{DR}
\corr{ricardo.vancini@duke.edu}{RV}
\corr{ikim9@uthsc.edu}{IK}
\corr{huang.tin@northeastern.edu}{TH}
\corr{o.vitek@northeastern.edu}{OV}
\corr{scott.soderling@duke.edu}{SHS}

% reduce space above and below equations
\setlength{\abovedisplayskip}{1pt}
\setlength{\belowdisplayskip}{3pt}


% main
\begin{document}

\maketitle


% abstract
\renewcommand{\abstractname}{Summary}
\begin{abstract}

In the review of this manuscript, significant concerns were raised by the
reviewers about the validity of our statistical approach to perform protein- and 
module-level inference from our \textbf{WASH-iBioID} and \textbf{SWIP-TMT} 
proteomics datasets. Our previous statistical approach was dependent upon the R 
package \texttt{edgeR} to evaluate differential protein abundance.
\texttt{edgeR} utilizes a negative binomial generalized linear
model (NB GLM) framework, originally developed for analysis of read counts data
generated in RNA-seq transcriptomics experiments.  Previously, we failed to fully
consider the validity of the NB GLM model used by \texttt{edgeR} for
proteomics data. In response to this critique, we explore the goodness-of-fit of
the NB GLM model for our \texttt{SWIP-TMT} data, and find evidence of a
lack-of-fit.  Thus, we revise our statistical approach and reanalyze
our data making use of the recently published tool \texttt{MSstatsTMT}.
\texttt{MSstatsTMT} uses a linear mixed-model (LMM) framework to model
major sources of variation in a proteomics experiment. We extend the LMM
framework used by \texttt{MSstatsTMT} to re-evaluate both protein- and
module-level statistical comparisions.  Despite evidence of a
lack-of-fit for the NB GLM method used by \texttt{edgeR}, we find that
the inferences we derived from our previous analysis are largely
preserved in our reanalysis using \texttt{MSstatsTMT}.
	
\end{abstract}


\newpage


\section{Lack-of-fit of the Negative Binomial Model}

Our previous approach is summarized as the 'Sum + IRS' method by Huang
\textit{et al.} (REF). Following protein summarization and Internal Reference
Scaling (IRS) normalization, we applied \texttt{edgeR} to assess differential
abundance of individual proteins and protein-groups or modules. 
The use of \texttt{edgeR} was based on work by Plubell 
and Khan, \textit{et al.} (REFS) who describe IRS normalization and the use of
\texttt{edgeR} for statistical testing in TMT mass spectrometry experiments. 
We failed however, to consider the overall adequacy of the NB GLM model for our
TMT proteomics data.\\

Statisitical inference in \texttt{edgeR} is performed for each gene or protein in
the dataset using a negative binomial framework in which the data are assumed to 
be adequately described by a NB distribution parameterized by a dispersion 
parameter, $\phi$. Practically, the dispersion parameter accounts for the
observed mean-variance relationship in proteomics and transcriptomics data.\\

%\texttt{edgeR} employs empirical Bayes methods that allow for the 
%estimation of feature-specific (i.e. gene or protein) biological variation, 
%even for experiments with small numbers of biological replicates, as is common 
%in transcriptomics and proteomics experiments. This empirical Bayes strategy is 
%a strength of the \texttt{edgeR} approach as it reduces the uncertainty of the 
%estimates and improves testing power.\\

As signal intensity in protein mass spectrometry is fundamentally related to the
number of ions generated from a ionized, fragmented protein, we incorrectly
inferred that TMT mass spectrometry data can be modeled as negative binomial
count data. Based on this assumption, we justified the use of \texttt{edgeR}.
Here, we reconsider the overall adequacy of the \texttt{edgeR} NB GLM model for
TMT mass spectrometry data.\\


\begin{figure}[hb!]
	\begin{fullwidth}
		\begin{center}
		\includegraphics[width=0.9\paperwidth,keepaspectratio]{gof}
		\caption{\textbf{Goodness-of-fit of \texttt{edgeR} (A), and 
		\texttt{MSstatsTMT} (B) statistical approaches.} The overall
		adequacy of the linear models fit to the data were assessed 
		by plotting the residual deviance for all proteins as a 
		quantile-quantile plot (McCarthy \textit{et al.}, (2012)). 
		\textbf{(A)} For analysis with \texttt{edgeR}, The normalized
		protein data from \texttt{MSstatsTMT} were fit with a negative
		binomal generalized linear model (NBGLM) of the form: 
		\texttt{Abundance} $\sim$ \texttt{Mixture + Condition}.
		Where \texttt{Mixure} is an additive blocking factor that 
		accounts for variablity between experiments. 
		The NB framework used by edgeR utilizes a dispersion parameter 
		to account for mean-variance relationships in the data.
		The dispersion parameter can take several forms. 
                \texttt{edgeR} supports three dispersion models: 'common',
		'trended', and 'tagwise'. However, when using \texttt{edgeR's}
		robust quasi-likelihood test methods, only global (i.e. 'common'
		or 'trended') dispersion metrics are appropriate 
		(see \texttt{edgeR::glmQLFit's} documentation). 
		We plot the protein-wise deviance from the data fit withe ach of
		the disperions parameters. Protein-wise deviance
		statistics were transformed to normality and plotted against
		theoretical normal quantiles using the \texttt{edgeR::gof}
		function. \textbf{(B)} For analysis with \texttt{MSstatsTMT},
		the normalized protein data were fit with a linear mixed-effects 
		model (LMM) of the form: 
		\texttt{Abundance} $\sim$ \texttt{0 + Condition + (1|Mixture)}. 
		Where \texttt{Mixture} represents the random effect
		of \texttt{Mixture}. The residual deviance and degrees of 
		freedom were extracted from the fitted models, z-score
		normalized, and plotted as in (A). Proteins with a significantly 
		poor fit are indicated as outliers in blue 
		(Holm-adjusted P-value $<$ 0.05).}
		\label{fig:gof}
	\end{center}
	\end{fullwidth}
\end{figure}


To evaluate the overall adequacy of the \texttt{edgeR} model, we plot the 
residual protein deviance statistics of all proteins against their theoretical, 
normal quantiles in a quantile-quantile (QQ) plot (\FIG{gof}).
The QQ plot addresses the question of how similar the observed data are to the
theoretical distribution given by a NB GLM fit. A linear relationship between the
observed and theoretical values is an indicator of goodness-of-fit.
Deviation from this linear trend is evidence of a lack-of-fit.\\

Following protein summarization and normalization with \texttt{MSstatsTMT}, the
data were fit with a simple NB GLM of the form \texttt{Abundance $\sim$ Mixture
+ Condition} using \texttt{edgeR's} \texttt{glmFit} function which fits a NB GLM
model to each protein or gene (the sub-subplot summaries) in the data. The
dispersion parameter $\phi$ can take several forms, and \texttt{edgeR} supports
three different dispersion metrics: 'common', 'trended', and 'tagwise'.
\FIG{gof} illustrates the divergence of the observed deviance
statistics from the theoretical distribution for our TMT data fit with the NB
GLM model.  These plots emphasize the overall lack-of-fit for proteomics data
with the \texttt{edgeR} model.\\ 

Given our experimental design, \texttt{MSstatsTMT} fits an analagous LMM:
\texttt{Abundance $\sim$ Condition + (1|Mixture)}. 
The QQ plot in \FIG{gof} indicates that the data are well 
described by \texttt{MSstatsTMT's} LMM framework, which 
does not depend upon the negative binomial assumption.


\section{Reanalysis of SWIP\textsuperscript{P1019R} TMT Proteomics}

Of note, most tools for analysis of protein mass spectromety data are derived
from tools originally developed for analysis of genomics and transcriptomics
data. An exception to this norm is \texttt{MSstatsTMT}, an extension of 
\texttt{MSstats} for analysis of TMT proteomics experiments.\\

\texttt{MSstatsTMT} utilizes a linear mixed-model framework. The strength of
LMMs lies in their flexibility. In mixed-models, the response variable is 
taken to be a function of both fixed- and mixed-effects. Using LMMs we can 
untangle the variance attributable to the biological effect we
are interested in from the experimental and biological covariates
which mask this response.\\

If the set of possible levels of the covariate is
fixed and reproducible then the factor is modeled as a fixed-effect parameter.
In contrast, if the levels of an observation reflect a sampling of the
set of all possible levels, then the covariate is modeled as a random effect.
Random or mixed-effects represent categorical variables
that reflect experimental or observational "units" in the data set (REF:Bates).
As such, mixed-effect parameters account for the variation occurring among all 
of the lower level units of a particular upper level unit in the data (REF:Bates).
For this reason, mixed-models may also be referred to as heirarchical models.\\

%Tandem mass tag, or TMT reagents enable the combination and simultaneous
%quantifiaction of multiple biological samples by mass spectrometry. Currently
%commercially available reagents are capable of labeling up to 16 protein
%preparations, which are then analyzed together in a single mass spectrometry run.
%Peptides labeled with TMT tags are distinguishable from each other due to the
%unique reporter ions generated by the TMT tag which  is used for relative
%quantification.  In a TMT experiment, ionized features are matched to peptides,
%these peptide spectrum matches (PSM), for all unique TMT channels are analyzed
%simultaneously as a single precursor. Quantification of all biological
%conditions is thus achieved within a single MS run in which all features for a
%protein are quantified simultaneously.\\

Huang \textit{et al.} created \texttt{MSstatsTMT}, an R package for data
normalization and hypothesis testing in multiplex TMT proteomics experiments. 
They outline a common vocabulary for describing the experimental design of 
TMT MS experiments. A TMT experiment consists
of the analysis of \texttt{m = 1} ... \texttt{M}\ concatenations of isobarically
labeled samples or \texttt{Mixtures}. This mixture is then analyzed by the mass
spectrometer in a mass spectrometry \texttt{Run}. This mixture is often
fractionated into multiple liquid chromotography \texttt{Fractions} to decrease
sample complexity, and thereby increase the depth of proteome coverage. 
Within a mixture, each of the unique TMT channels is dedicated to the 
analysis of \texttt{c = 1} ... \texttt{C}\ individual biological or treatment 
\texttt{Conditions}.  There may then be \texttt{b = 1} or more \texttt{B}\ 
biological replicates or \texttt{Subjects}. Finally, a single TMT mixture may be 
repeatedly analyzed in \texttt{t = 1} ... \texttt{T}\ technical replicate mass 
spectrometry runs.\\

\textbf{Equation} is a mixed-effects formula which describes a general TMT
experiment composed of \texttt{M} mixtures, \texttt{T} technical replicates of 
mixture, \texttt{C} conditions, and \texttt{B} biological subjects.
The abundance of a given protein, $Y_{mcbt}$, is then:\\

\begin{equation}
	% full model
	Y_{mcbt} = \mu + Mixture_m + TechRep(Mixture)_{m(t)} + Condition_c + 
	Subject_b + \epsilon_{mcbt}
\end{equation}

The model's constraints distinguish fixed and random components of 
variation in the response.\\

\begin{equation}
  \begin{gathered}
	% constraints:
	Mixture_m \stackrel{iid}{\sim} N(0,\sigma^2_M) \\
	TechRep(Mixture)_{t(m)} \stackrel{iid}{\sim} N(0,\sigma^2_T) \\
	\sum_{c=1}^{C} Condition_c = 0 \\
	Subject_{mcb} \stackrel{iid}{\sim} N(0,\sigma^2_S) \\
	\epsilon{mtcb} \stackrel{iid}{\sim} N(0,\sigma^2) \\
  \end{gathered}
\end{equation}

\texttt{Mixture} is a mixed-effect and .
represents the variation between TMT mixtures. By definition mixed-effects are 
assumed to be independent and normally distributed (iid). 
\texttt{TechRep(Mixture)} represents random 
variation between replicate mass spectrometry runs.
The term \texttt{Subject} cooresponds to each unique biological replicate and 
represents biological variation among the levels of the fixed-effect term
\texttt{Condition}. The term $\epsilon_{mtcb}$,
is a random-effect representing both biological and technical variation, 
quantifying any remaining error.\\

If a component of the model is not estimable, then it is removed. 
For example, if there is no technical replication of mixture 
\texttt{(T=0)}, the model is reduced to: \\

\begin{equation} 
	% reduced model (fx0)
	Y_{mcbt} = \mu + Mixture_m + Condition_c + \epsilon_{mcb}
\end{equation}

In the reduced model, biological variation among individual \texttt{Subjects} is 
captured by the term \texttt{Condition}, and is thus omitted\\


\section{Hypothesis Testing}

\texttt{MSstatsTMT} performs protein-wise
comparisons between \texttt{Conditions} of biological \texttt{Subjects} by 
a contrast of conditioned means obtained from fitting the data with a
linear mixed-effects model expressing the major sources of variation in the
experimental design.\\

Model-based testing of differential abundance between pairs of conditions
is done by comparing the estimates obtained from the fit LMM. We are interested 
in testing the null hypothesis $H0 : l^T\beta = 0$. Kutzenova \textit{et al.,}
derive a test statistic for such contrasts (Kutzenova2017):

\begin{equation}
	% t-statistic
	t = \frac{l^T \hat{\beta}}{\sqrt{l \sigma^2 \hat{V} l^T}}
\end{equation}

We obtain the model estimates $\hat{\beta}$, 
error $\sigma^2$, and
variance-covariance matrix $\hat{V}$ from the fitted model. 
Together $\sigma^2 * \hat{V}$ is the scaled variance-covariance matrix
describing the error estimates of the model's mixed-effect parameters. 
Given $l^T$, a vector of sum 1 specifying the positive and negative
coefficients of the comparison, the numerator of the equation is then
the fold change of a given comparison, and together the denominator 
represents the standard error of the contrast.\\

The degrees of freedom for the contrast are derived using the Satterthwaite
moment of approximation method (Kutzenova2017).
Finally, given the t-statistic, which is assumed to follow an approximate
$\chi^2$ distribution, and the degrees of freedom, a p-value is calculated. 
P-values for the protein-wise tests are adjusted using the
Benjamini-Hochberg (REF).\\


\begin{figure}[ht!] %% figure 2 -- Experimental Design
  \begin{fullwidth}
  \begin{center}
	  \includegraphics[width=0.9\paperwidth,keepaspectratio]{design}
	  \caption{\textbf{Experimental Design.} We performed three 16-plex TMT
	  experiments. Each TMT mixture is a concatenation of 16 labeled
	  samples. In each experiment we analyzed 7 subcellular
	  \texttt{BioFractions} prepared from the brain of a 'Control' or
	  'Mutant' mouse. In all we analyzed 3 \texttt{Subjects} from each 
	  {Condition}. Each \texttt{Mixture} includes two \texttt{Channels}
	  dedicated to the analysis of a common quality control sample.}
	  \label{fig:design}
  \end{center}
  \end{fullwidth}
\end{figure}


\section{SWIP-TMT Proteomics Experimental Design}

In our experiment, the fixed-effect term \texttt{Condition} represents the 14
combinations of \texttt{Genotype} and \texttt{BioFraction} obtained from 
subcellular fractionation of the brains of 'Control' and 
SWIP\textsuperscript{P1019R} 'Mutant' mice. We refer to these as
\texttt{BioFractions} to distinguish them from a MS \texttt{Fraction}. 
Our TMT proteomics experimental design is summarized in \FIG{design}.

In our experiment, each TMT mixture contains seven repeated measurements
made from each biological \texttt{Subject}. To account for this source of
intra-Subject variability, we should include the random-effect term
\texttt{Subject} representing the random error within a subject. However, in our
design \texttt{Mixture} is confounded with the term \texttt{Subject}. In
each mixture we analyzed all \texttt{BioFractions} from a single Control and
Mutant mouse.  Thus we can choose to account for the effect of \texttt{Mixture}
or \texttt{Subject}, but not both. We choose to account for variability of 
\texttt{Mixture} under the assumption that the effect of this experimental batch 
effect is greater than the variance attributable to the random variability
inherent to repeated measurements of each subject. 
Thus we omit the term \texttt{Subject}. The reduced model is then the same as
equation (EQ) when \texttt{Condition} is the interaction of 
\texttt{Genotype:BioFraction}.\\


\section{Protein level comparisions}

Using \texttt{MSstatsTMT} we assesssed two protein-level comparisons:

\begin{itemize}
	\item 'intra-BioFraction' contrasts 
	\item 'Mutant-Control' contrast 
\end{itemize}

\begin{figure}[h!] %% figure 3 -- Contrasts
  \begin{fullwidth}
  \begin{center}
	  \includegraphics[width=0.9\paperwidth,keepaspectratio]{contrasts}
	  \caption{\textbf{Statistical Comparisons.} We assessed two types of
	  contrasts. Each row of the matrix specifies a contrast between
	  positive and negative coefficients in the mixed-effects model fit to
	  each protein. Contrasts1-7 are 'intra-BioFraction' contrasts that
	  specify the pairwise comparisons of Control and Mutant groups for a
	  single fraction. In Contrast 8 we compare 'Mutant-Control' and asses
	  the overall difference of 'Control' and 'Mutant' conditions.  Each
	  contrast is a vector of sum 1.}
	  \label{fig:contrasts}
  \end{center}
  \end{fullwidth}
\end{figure}


'Intra-BioFraction' comparisons are the 7 pairwise comparisons of 'Control' and
'Mutant' protein abundance for each subcellular fractions. We
also assessed differential abundance for the overall comparison between 'Control'
and 'Mutant' groups. Each of these contrasts is represented by a vector, $l^T$, 
which specifies a comparison between coefficients in the LMM. \FIG{contrasts}
illustrates a matrix defining all 8 contrasts. 

\texttt{MSstatsTMT} attempts to automatically parse the experimental design and
fit an appropriate LMM for the experimental design.
In order to understand and extend the function of \texttt{MSstatsTMT}, we 
extracted \texttt{MSstatsTMT's} core model-fitting and statistical testing 
steps and illustrate them here.\\

Following data preprocessing, summarization, and normalization, statistical
inference by \texttt{MSstatsTMT} can be summarized in two steps:\\

\begin{itemize}
	\item Fit each protein with the appropriate LMM, and then
	\item given the fitted model, assess a contrast of interest.
\end{itemize}
		
At the core of the model fitting-step is the
R package \texttt{lme4} which implements mixed-effects models with its function
\texttt{lmer}. The package \texttt{lmerTest} extends \texttt{lme4's}
functionality and enables the computation of Sattertwaite degrees of freedom.\\

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<fit0, eval=TRUE, echo=TRUE, results='hide'>>=

# load dependencies
library(dplyr)
library(data.table)

#library(SwipProteomics)

# load the data
data(swip)
data(msstats_prot)

# formula to be fit to WASHC4, aka SWIP:
fx <- formula("Abundance ~ 0 + Genotype:BioFraction + (1|Mixture)")

# fit the LMM
fm <- lmerTest::lmer(fx, msstats_prot %>% filter(Protein == swip))

# examine the model's summary
summary(fm, ddf = "Satterthwaite")

@

<<summary, eval=TRUE, echo=FALSE>>=

df <- summary(fm, ddf = "Satterthwaite")$coefficients %>% 
        as.data.table(keep.rownames="Coefficient") %>%
	mutate(Coefficient = gsub("Genotype|BioFraction","",Coefficient)) %>% 
	mutate("p value" = formatC(`Pr(>|t|)`, digits=3)) 
df$`Pr(>|t|)` <- NULL
df %>% knitr::kable()
message("\n")

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

The model's estimates, $\beta$, represent our best estimate of the mean protein
abundance in the 14 conditions of \texttt{Genotype:BioFraction}. To illustrate a
comparison, we define a contrast comparing 'Mutant:F7' and 'Control:F7'.
The function  \texttt{lmerTestContrast} performs model-based comparisons of
conditions defined by a contrast matrix. While the work done by this function 
is the same as the work done internally by \texttt{MSstatsTMT's}
\texttt{groupComparisonsTMT} function, \texttt{lmerTestContrast} is more
flexible.\\

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<contrast0, eval=TRUE, echo=TRUE, results='hide'>>=

# create a contrast
coeff <- lme4::fixef(fm)
contrast7 <- setNames(rep(0,length(coeff)), nm = names(coeff))
contrast7["GenotypeMutant:BioFractionF7"] <- +1 # positive coeff
contrast7["GenotypeControl:BioFractionF7"] <- -1 # negative coeff

# evaluate contrast
lmerTestContrast(fm, contrast7)

@

<<pretty-print, eval=TRUE, echo=FALSE>>=

df <- lmerTestContrast(fm, contrast7) 
df <- df %>% mutate(Contrast = gsub("Genotype|BioFraction","",Contrast))  
df$SE <- round(df$SE,3)
df$log2FC <- round(df$log2FC,3)
df$percentControl <- NULL
df$Tstatistic <- round(df$Tstatistic,3)
df$Pvalue <- formatC(df$Pvalue,digits=3)
df$isSingular <- NULL
df %>% knitr::kable()
message("\n")

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Provided the correct contrast, we easily assess the overall comparison between 
'Mutant' and 'Control' groups:

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<<contrast1, eval = TRUE, echo = TRUE, results='hide'>>=

# use convenience function to contruct a contrast
lT <- getContrast(fm, "Mutant","Control")

# assess the comparison
lmerTestContrast(fm, lT)

@

<<<pretty-print results, eval = TRUE, echo = FALSE>>=

df <- lmerTestContrast(fm, lT)
df <- df %>% mutate(Contrast = 'Mutant-Control')
df$SE <- round(df$SE,3)
df$log2FC <- round(df$log2FC,3)
df$percentControl <- NULL
df$Tstatistic <- round(df$Tstatistic,3)
df$Pvalue <- formatC(df$Pvalue,digits=3)
df$isSingular <- NULL
df %>% unique() %>% knitr::kable()

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\section{Goodness-of-fit}

It is useful to consider the goodness-of-fit of our LMM. A straight forward
measure of a LMM's quality is the Nakagawa coefficient of 
determination (REF). Nakagawa's conditional $R^2$ is interpreted as 
the total variance explained by a LMM ($R^2_{total}$).
The marginal $R^2$ is interpreted as the variance explained by the LMM's 
fixed-effects ($R^2_{fixed}$).

We implment Nakagawa's coeffficient of determination using the 
\texttt{r.squaredGLMM.merMod} function taken from the \texttt{MuMin}
package (REF).\\


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<rsquared, eval=TRUE, echo=TRUE>>=

# assess gof with Nakagawa coefficient of determination
r.squaredGLMM.merMod(fm)

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We can see the total variation explained by the model, $R^2_{c}$, is 0.949. 
The variance explained by fixed-effects, \texttt{Genotype:BioFraction},
equates to 0.935 ($R^2{m}$). A vast majority of the variance is attributable to 
the fixed-effects. Only about 1.5\% of the remaing variance is attributable to 
mixed-effects and the residual variance.\\

\section{Module-level analysis}

We wish to extend the LMM framework developed by \texttt{MSstatsTMT} to perform 
inference at the level of protein groups or modules.
That is, for module-level comparisons, we are interested in the overall affect 
of Genotype on a group of proteins. Where modules are groups of covarying 
proteins which represent biological niches defined by proteins that 
localized together in subcellular space.\\

Here we hypothesize that the proteins within a module, which are a subset of 
the overall proteome,  are a part of a common group, a module, with a common 
mean effect. Proteins within a module are correlated observations 
which we model as a mixed-effect. 
We take the stance that \texttt{Protein} is a mixed-effect 
in the view that we are primarily interested in making inference about the 
overall distribution of responses for a module rather than among its sublevels.\\

We model protein groups or modules by adding the mixed-effect term
\texttt{Protein} to the LMM equation (EQ):

\begin{equation} 
  \begin{gathered}
	Y_{mcbt} = \mu + Mixture_m + Condition_c + Protein_p + \epsilon_{mcb}\\
	Protein{p} \stackrel{iid}{\sim} N(0,\sigma^2_P) \\
  \end{gathered}
\end{equation}

The term \texttt{Protein} is associated with a variance component $\sigma_P$.\\

As a means of example, we demonstrate an ideal module, by fitting the LMM to 
the 5 WASH complex proteins. As before, we calculate the coefficient of 
determination for LMM's with the \texttt{r.squaredGLMM.merMod} function
(REF).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<module model, eval=TRUE, echo=TRUE>>=

# the module-level formula to be fit:
fx <- "Abundance ~0 + Genotype:BioFraction + (1|Mixture) + (1|Protein)"

# load WASH Complex proteins
data(washc_prots)

fit <- lmer(fx, msstats_prot %>% filter(Protein %in% washc_prots))

r.squaredGLMM.merMod(fit)

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Again, we consider the variance explained by the model as a measure of its
overall quality. Our model explains 0.89\% of the total variance.
The Fixed-effect term \texttt{Genotype:BioFraction} explains the majority of 
the variance ($R^2_m=0.762$). The remaining variance, 0.13\%, is attributable to
the combination of mixed-effects \texttt{Mixture} and \texttt{Protein} as well as the residual variance.\\

It is useful to consider the variation ($\sigma^2$) of the 
individual mixed-effect terms. These can be accessed with \texttt{lme4's} 
\texttt{VarCorr} function.\\

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<VarCorr, eval=TRUE, echo=TRUE>>=

# calculate variance of mixed-effects
var_df <- as.data.frame(lme4::VarCorr(fit,comp="Variance"))
mixef_var <- setNames(var_df$vcov,nm=var_df$grp) # var of each component
mixef_var/sum(mixef_var) # as a percent of the total mixed-effect variance

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\begin{figure}[h!] %% figure 4 -- Variance Partition
  \begin{fullwidth}
  \begin{center}
	  \includegraphics[width=0.9\paperwidth,keepaspectratio]{variance}
	  \caption{\textbf{Analysis of Variance Components.} 
	  The proportion of variance explained by Genotype, BioFraction,
	  Mixture, and remaining residual error (subplot error) for all
	  proteins. Note while the contribution of Mixture seems negligiable,
	  its average for all proteins is approximately twice the average
	  percent variance explained by Genotype. BioFraction explains the
	  majority of the variance for all proteins. Analysis done with
	  \texttt{variancePartition::calcVarPart}.}
	  \label{fig:variance}
  \end{center}
  \end{fullwidth}
\end{figure}


The R package \texttt{variancePartition} enables us to calculate the percent
variance explained by a LMM's parameters. To do so, it expects all terms to be
mixed-effects. \FIG{variance}.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<variancePartition, eval=TRUE, echo=TRUE>>=

# load variancePartition
library(variancePartition)

# calculate partitioned variance
fx <- "Abundance ~ (1|Genotype) + (1|BioFraction) + (1|Mixture) + (1|Protein)"
fit <- lmer(fx, data = msstats_prot %>% filter(Protein %in% washc_prots))

calcVarPart(fit)

@
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


We can see that the majority of the variance explained by the LMM fit to the
WASH complex is attributable to \texttt{Genotype}. The mixed-effect terms
\texttt{Protein} and \texttt{Mixture} account for a small fraction of the 
overall variance explained by the model.\\

As our overall goal is to identify groups or modules of proteins that strongly
covary together, our clustering approach should maximize the variance explained
by a modules fixed-effect parameters (Genotype + BioFraction) while minimizing 
the variance among its individual proteins. 
An ideal module is a perfect summary of its protein constituents, 
$PVE_{Protein}=0$. We use this idea of a module's quality to suprivise our 
clustering approach.\\

\begin{equation}
	Quality_{Module}=\frac{PVE_{Genotype} + PVE_{BioFraction}}{PVE_{Protein}}
\end{equation}


\section{Network Construction}

Using our \textbf{SWIP-TMT} dataset, we aim to identify modules or groups of
proteins that covary together across subcellular space. Prior to building the
co-variation network, other sources of variation should be removed. Although
\texttt{MSstatsTMT} handles the batch effect inherent in experiments with
multiple TMT mixtures, it is necessary to remove this effect prior to building
the network. We removed the effect of \texttt{Mixture} using
\texttt{limma::RemoveBatchEffect}.  These adjusted data are used for network
construction and plotting but not statistical modeling.\\

Prior to network construction, we removed protein models with poor fit 
($R^2_{total}<0.7$; n=791 proteins). Removing this noisey proteins facilitation
module identification and improves overall module quality.\\

The final network was constructed using data from both 'Control' and 'Mutant' 
samples after adjusting for batch (Mixture). The final dataset included 
42 samples and 6,119 proteins. The protein covariation network was build by
calculating the Pearson correlation for all pairwise comparisons of proteins.\\

We performed network enhancement to remove biological noise from the network.
This step is essential for module detection. Network enhancement reweights the
network's edges and has the overall effect of making the network sparse.
Conceptually this step is related to the soft-thresholding approach taken by 
WGCNA or WPCNA analysis workflows (REFS), but has the befinit of not assuming
that the network has an overall scale free topology. 
Without reweighting or enhancing the network, most extant clustering 
algorithms fail to detect communities in the dataset. 
Network enhancment has the effect of making the
network sparse and facilitates the identification of network structure (FIG).\\

\end{document}
