% title: response.tex
% description: Response to eLife Reviewers
% author: twab

% USAGE: 
% to compile this document:
%    R  >>> knitr::knit("response.Rnw")
%    sh >>> pdflatex response.tex


<<renv, echo=FALSE, cache=FALSE, results='hide'>>=

# prepare the R environment
root <- "~/projects/SwipProteomics"
renv::load(root,quiet=TRUE)
devtools::load_all(quiet=TRUE)

suppressPackageStartupMessages({
  library(knitr)
  library(dplyr)
  library(data.table)
})

# global knitr options
opts <- list(tidy=FALSE,message=FALSE,warning=FALSE,fig.width=4,fig.height=4.5)
do.call(opts_chunk$set,opts)

@


% document setup

\documentclass[11pt]{elife}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsthm}
\usepackage{ragged2e}
\usepackage{caption}
\usepackage{fancyhdr}
\usepackage{graphicx}
\usepackage{titlesec}
\usepackage{blkarray}
\usepackage{csquotes}

% location of figures for this compiled document
\graphicspath{ {./figs/} }


\title{Supplementary Methods and Response to eLife Reviewers\\
\small{Genetic Disruption of WASHC4 Drives Endo-lysosomal Dysfunction and \\
Cognitive-Movement Impairments in Mice and Humans}}


% Authors
\author[1\authfn{0}]{Jamie Courtland}
\author[1\authfn{0}]{Tyler W. A. Bradshaw}
\author[2]{Greg Waitt}
\author[2,3]{Erik J. Soderblom}
\author[2]{Tricia Ho}
\author[4]{Anna Rajab}
\author[5]{Ricardo Vancini}
\author[2\authfn{1}]{Il Hwan Kim}
\author[6]{Ting Huang}
\author[6]{Olga Vitek}
\author[3]{Scott H. Soderling}

\affil[1]{Department of Neurobiology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[2]{Proteomics and Metabolomics Shared Resource, 
Duke University School of Medicine, Durham, NC 27710, USA}
\affil[3]{Department of Cell Biology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[4]{Burjeel Hospital, VPS Healthcare, Muscat, Oman}
\affil[5]{Department of Pathology, Duke University School of Medicine, 
Durham, NC 27710, USA}
\affil[6]{Khoury College of Computer Sciences, Northeaster University,
Boston, MA 02115, USA}

\contrib[\authfn{0}]{These authors contributed equally to this work.}
\presentadd[\authfn{1}]{Department of Anatomy and Neurobiology, 
University of Tennessee Health Science Center, Memphis, TN 38163, USA}

\corr{jlc123@duke.edu}{JC}
\corr{tyler.w.bradshaw@duke.edu}{TWAB}
\corr{greg.waitt@duke.edu}{GW}
\corr{erik.soderblom@duke.edu}{EJB}
\corr{tricia.ho@duke.edu}{TH}
\corr{drannarajab@gmail.com}{DR}
\corr{ricardo.vancini@duke.edu}{RV}
\corr{ikim9@uthsc.edu}{IK}
\corr{huang.tin@northeastern.edu}{TH}
\corr{o.vitek@northeastern.edu}{OV}
\corr{scott.soderling@duke.edu}{SHS}

% page header
\rhead{Response to eLife Reviewers} 

% reduce space above and below equations
\setlength{\abovedisplayskip}{3pt}
\setlength{\belowdisplayskip}{3pt}


\begin{document}

\maketitle


\begin{abstract}

In the review of this manuscript significant concerns were raised by the
reviewers about the validity of our statistal approach to perform protein- and 
module-level inference from our \texttt{SWIP-TMT} proteomics dataset. 
Here we address these concerns and provide additional supplemental description of out statistical methods.

\end{abstract}

\newpage


\section{Reanalysis of SWIP\textsuperscript{P1019R} TMT Proteomics}

Previously, we analyzed both the \texttt{WASH-BioID} and \textt{SWIP-TMT}
proteomics datasets with the \texttt{edgeR} package for R which was originally
designed for analysis of read counts generated by RNA-seq or similar 
technologies. 

Several parallels between proteomisc and transcriptomics data can be drawn.
Of note, most extant analytical tools for analysis of proteomics data are 
derived from tools originally designed for analysis of genomic or transcriptomics
datasets. Moreover, signal intensity in proteomics experiments is related to the 
number of ions quantified for a given ionized species, and from this idea arises
the confusion that feature signal intensity may be modeled as a NB count matrix.

Additionally, inference in the transcriptomics and proteomisc is performed in
two steps, gene or protein summarization and subsequent statistical testing on
the summarized data. 

edgeR employs empirical Bayes methods that allow for the estimation of 
gene-specific biological variation, even for experiments with
minimal levels of biological replication.
As in proteomics analysis, edgeR is applied to assess differential expression at
the level of summarized sub-subplots (genes or proteins).

We preprocessed peptide level data exported from \texttt{Proteome Discoverer}
using custom R scripts and then performed statistical interence at the level of 
proteins and protein groups or modules using \texttt{edgeR's} \texttt{exactTest}
or \texttt{glmQLF} functions. \\ 

Statistical inference in \texttt{edgeR} is built on a 
negative binomial (NB) model framework. The data are assumed to be 
adequately described by a NB distribution parameterized by a dispersion 
parameter, $\phi$. 

NBdisp = global param
QLdist = gene-specific variability


The NB dispersion describes the overall biological variability across all genes. 

edgeR extends the NB model with quasi-likelihood (QL) methods to account for
gene or protein specific variability from both biological and technical sources 
[edgeRuserGuide:23, 22]. 
Under the QL framework,
the variance of the count ygi is a quadratic function of the mean,
var(ygi) =  T);

where  is the NB dispersion parameter and he dispersion paramer empirical Bayes
estimate of the negative binomial dispersion parameter for each s the QL
dispersion parameter.  Any increase in the observed variance of ygi will be
modelled by an increase in the estimates for  and/or protein or gene
In this model, the NB dispersion  is a global parameter whereas the QL is
gene-specific, so the two dispersion parameters have different roles. 


The QL dispersion picks up any gene-specific variability above and below the
overall level.

The exact test is based on the qCML methods. Knowing the conditional
distribution for the sum of counts in a group, we can compute exact p-values by
summing over all sums of counts that have a probability less than the
probability under the null hypothesis of the observed sum of counts. The exact
test for the negative binomial distribution has strong parallels with Fisher’s
exact test.

QL F-test is preferred as it reflects the uncertainty in estimating the
dispersion for each gene. It provides more robust and reliable error rate
control when the number of replicates is small.

In practice, we use the trended dispersions to account for the empirical
mean-variance relationships.
Since the NB dispersion under the QL framework reflects the overall biological
variability

Estimation of the gene-specific QL dispersion is difficult as most RNA-seq data
sets have limited numbers of replicates. This means that there is often little
information to stably estimate the dispersion for each gene. To overcome this,
an empirical Bayes (EB) approach is used whereby information is shared between
genes [38, 23, 28]. Briefly, a mean-dependent trend is fitted to the raw QL
dispersion estimates. The raw estimates are then squeezed towards this trend to
obtain moderated EB estimates, which can be used in place of the raw values for
downstream hypothesis testing. This EB strategy reduces the uncertainty of the
estimates and improves testing power.o

The dispersion parameter can take several forms. 
\texttt{edgeR} supports three dispersion models: 'common',
'trended', and 'tagwise'. However, when using \texttt{edgeR's}
robust quasi-likelihood test methods, only global (i.e. 'common'
or 'trended') dispersion metrics are appropriate 
(see \texttt{edgeR::glmQLFit's} documentation).

At the core of our normalization appraoch is the use of common quality control
(QC) samples, analyzed in technical duplicate in each TMT mixture. The sample 
pool QC sample was formed by combining aliquots of all biological replicates.
This sample therefore is representative biological and technical variation.
We utilzize the IRS or internal reference standard normalization described
by Plubell et al., 2017.

In practice, we find that it is essential to insure that the input data for
clustering is of the highest quality in order too...

MSstatsTMT performs normalization within each TMT Run or Mixture:Channel. 
We remove PSMs that exhibit abnormal variablity, accounting for this
mean-variance relationsihp.

The most important step in our normalization approach is 
IRS normalization. MS2 random sampling results in identification and
quantification of proteins by different peptides in each MS experiment.
To account for this source of variability, protein measurements are
adjusted by a scaling factor such that the geometric mean of all
internal reference standards are equal (Plubell et al., 2017).
This is essential to account for the stochasticisity of peptide
quantification in MS experiments. 

Our previous appraoch can be sumamrized as the "Sum + IRS Normalization" method
described by Huang et al., 2020. We summarize proteins as the sum of its
features and use IRS noramlization. A potential weakness of the sum method is
that outlier peptides can strongly influence the summary value. Median-based
methods (median and median-polish) avoid this problem. We examined outliers in a
method described by Ping et al., (20xx).


%% figure 1 
\begin{figure}[ht]
	\begin{fullwidth}
		\begin{center}
		\includegraphics[width=0.9\paperwidth,keepaspectratio]{gof}
		\caption{\textbf{Goodness-of-fit of \texttt{edgeR} (A), and 
		\texttt{MSstatsTMT} (B) statistical approaches.} The overall
		adequacy of the linear models fit to the data were assessed 
		by plotting the residual deviance for all proteins as a 
		quantile-quantile plot (McCarthy \textit{et al.}, (2012)). 
		\textbf{(A)} For analysis with \textt{edgeR}, The normalized
		protein data from \texttt{MSstatsTMT} were fit with a negative
		binomal generalized linear model (NBGLM) of the form: 
		\texttt{Abundance} $\sim$ \texttt{Mixture + Condition}.
		Where \texttt{Mixure} is an additive blocking factor that accounts for
		variablity between experiments. Protein-wise deviance
		statistics were transformed to normality and plotted against
		theoretical normal quantiles using the \texttt{edgeR::gof}
		function. \textbf{(B)} For analysis with \texttt{MSstatsTMT},
		the normalized protein data were fit with a linear mixed-effects 
		model (LMM) of the form: 
		\texttt{Abundance} $\sim$ \texttt{0 + Condition + (1|Mixture)}. 
		Where \texttt{Mixture} represents the random effect
		of \texttt{Mixture}. The residual deviance and degrees of 
		freedom were extracted from the fitted models, z-score
		normalized, and plotted as in (A). Proteins with a significantly 
		poor fit are indicated as outliers in blue 
		(Holm-adjusted P-value $<$ 0.05).}
		\label{fig:gof}
	\end{center}
	\end{fullwidth}
\end{figure}

We evaluated the overall adequacy of the \texttt{edgeR} model by plotting the 
residual deviance of all proteins against their theoretical, 
normal quantiles in a quantile-quantile plot. \textbf{Figure \ref{fig:gof}} 
illustrates the overall lack of fit for the three disperion models fit by 
\texttt{edgeR}.  As an alternative to \texttt{edgeR} we considered 
\texttt{MSstatsTMT}, an extension of MSstats for analysis of TMT proteomics
experiments. \\

\texttt{MSstatsTMT} utilizes a linear mixed-model framework. The strength of 
linear mixed models (LMMs) is in their ability to account for complex sources of 
variation in an experimental design. 

Mixed models use both fixed and random effects. These correspond
to a hierarchy of levels with the repeated, correlated measurement
occurring among all of the lower level units for each particular upper
level unit.

\begin{displayquote}
	In a mixed model one or more covariates 
	are a categorical variable  representing experimental or 
	observational "units" in the data set. 
	[...]
	If the set of possible levels of the covariate is fixed and reproducible 
	we model the covariate using fixed-effects parameters. 
	If the levels that we observed represent a random sample from
	the set of all possible levels we incorporate random effects in the
	model.
\end{displayquote}

Huang et al., describe a general statistical framework for TMT mass spectrometry
experiments. A TMT proteomics experiment consists of the analysis of 
\texttt{m =1} ... \texttt{M} concatensions of isobarically labeled samples or \texttt{Mixtures}. 
Within a mixture, each TMT channel is dedicated to the analysis of
\texttt{c = 1} ... \texttt{C} individual biological or treatment \texttt{Conditions} 
prepared from \texttt{b = 1} ... \texttt{B} biological replicates or \texttt{Subjects}. 
A single mixture may be profiled in \texttt{t = 1} ... \texttt{T} technical replicate 
mass spectrometry runs. 

The following mixed-effects model describes protein abundance measured in an
in an experiment composed of M mixtures, T technical replicates of Mixture, C
conditions, and B biological Subjects. 

%% Full model
\begin{equation} \label{eq1}
	Y_{mcbt} = \mu + Mixture_m + TechRep(Mixture)_{m(t)} + Condition_c + \epsilon_{mcbt}
\end{equation}

Mixture a variation between
mixtures, and 
nd TechRep(Mixture) distinguish technical
between replicate mass spectrometry runs of a same mixtureo

If a component of the model is not estimable, it is removed. Thus, if there is
no technical replication of Mixture, the model is reduced to: \\

%% Reduced model - no technical replication of mixture
\begin{equation} \label{eq2}
	Y_{mcbt} = \mu + Mixture_m + Condition_c + Subject_{mcb} + \epsilon_{mcb}
\end{equation}

Control and Mutant mouse.

It is assumed that error at the level of protein summaries are independent and
non-systematic (no mean variance) errors at the level of protein summaries εmtcb can be assumed
independent and non-systematic.

Where Mixture is a mixed-effect and represents variation between TMT mixtures.
Subject is a mixed-effect and represents variation between biological Subjects
Condition is a fixed effect. The term $\epsilon$ 
is a random effect representing both biological and technical variation, 
quantifying any remaining error ($\sigma^2$). \\

The constraints distinguish the fixed and
the random components of variation. Comparisons of interest (i.e., conditions),
are at the level of the subplot. Their
standard errors are derived from the subplot components of variatio
Subject represetnts biological replicates using the convention taht each
biological replicate has a unique identifier across Mixtures and Conditions. 


To distinguish between Mixtures which may be fractionated, and these
\texttt{Fractions} analyzed to increase the depth of coverage in an MS
experiment, we refer to fractions as \texttt{BioFraction}.

In our experiment, the fixed effect term Condition 
represents interaction of Genotype and BioFraction--the 14 combinations 
of 7 BioFractions fractions from Control and Mutant mice. 

We prepared 7 \texttt{BioFractions} from 'Control' and SWIP\textsuperscript{P1019R} 'Mutant' mice. 
14 conditions
We analyzed data from 6 Subjects, three bioreplicate Control and 
SWIP\textsuperscript{P1019R} Mutant mice.

%% reduced model - no repeated measures of Subject
\begin{equation} \label{eq3}
	Y_{mcbt} = \mu + Mixture_m + Condition_c + \epsilon_{mc}
\end{equation}




\begin{figure}[h]
  \begin{fullwidth}
  \begin{center}
	  \includegraphics[width=0.9\paperwidth,keepaspectratio]{design}
	  \caption{\textbf{Experimental Design.} We performed three 16-plex TMT
	  experiments. Each TMT mixture is a concatenation of 16 labeled
	  samples. In each experiment we analyzed 7 subcellular
	  \texttt{BioFractions} prepared from the brain of a 'Control' or
	  'Mutant' mouse. In all we analyzed 3 \texttt{Subjects} from each 
	  {Condition}. Each \texttt{Mixture} includes two \txttt{Channels}
	  dedicated to the analysis of a common quality control sample.}
	  \label{fig:design}
  \end{center}
  \end{fullwidth}
\end{figure}

In our experimental design, we made measurments from seven BioFractions 
from each subject. Thus, we should include the term 
Subject, representing the 6 individual mice or subjects analyzed in our 
experiment. \\


\begin{figure}[h]
  \begin{fullwidth}
  \begin{center}
	  \includegraphics[width=0.9\paperwidth,keepaspectratio]{contrasts}
	  \caption{\textbf{Statistical Comparisons.} We assessed two types of
	  contrasts. Each row of the matrix specifies a contrast between
	  positive and negative coefficients in the mixed effects model fit to
	  each protein. Contrasts1-7 are 'intra-BioFraction' contrasts that
	  specify the pairwise comparisons of Control and Mutant groups for a
	  single fraction. In Contrast 8 we compare 'Mutant-Control' and asses
	  the overall difference of 'Control' and 'Mutant' conditions.  Each
	  contrast is a vector of sum 1.}
	  \label{fig:contrasts}
  \end{center}
  \end{fullwidth}
\end{figure}

However, in our design \texttt{Mixture} is confounded with the
term \texttt{Subject} -- in each mixture we analyzed all \texttt{BioFractions} 
from a single Control and Mutant mouse. Thus we can choose to account for the 
effect of \texttt{Mixture} or \texttt{Subject}, but not both. 
Assuming Mixture contributes greater to the variance, we drop the term Subject,
and the reduced model is equivalent  to \ref{eq1}. \\

Model based testing of differential abundance between pairs of conditions
is assessed through contrast of conditioned means estimated by fitting the
parameters of the model by REML to obtain $\hat{\beta}$, $\sigma^2$ and
$\hat{V}$.

The degrees of freedom are determined by the Satterthwaite approximation[REF],
and the T-statistic for the contrast is taken to be (lmerTest ref): \\

%% T-statistic Equation
\begin{equation}
	t = \frac{l^T * \hat{\beta}}{sqrt(l * \sigma^2 * \hat{V} * l^T)}
\end{equation}

$\sigma^2$ is the error from \textbf{Equation} \ref{eq1}.
$l^T$ is a vector specifying a contrast between positive and 
negative coefficients in the model.

Together, the denominator $\sqrt(l * \sigma^2 * \hat{V} * l^T)$ is the 
standard error of the contrast.


<<fit the model, eval=TRUE, echo=TRUE>>=

suppressPackageStartupMessages({
  library(dplyr)
  library(data.table)
})

## load SwipProteomics data
data(swip)
data(gene_map)
data(msstats_prot)
data(alt_contrast)
data(msstats_contrasts)

## formula to be fit:
fx0 <- formula("Abundance ~ 0 + Condition + (1|Mixture)")

# fit the model
idx <- msstats_prot$Protein == swip
fm <- lmerTest::lmer(fx0, msstats_prot[idx,])

# calculate model statistics
model_summary <- summary(fm,ddf="Satterthwaite")

df <- model_summary$coefficients
df %>% as.data.table(keep.rownames="Coefficient") %>% knitr::kable()

# evaluate goodness-of-fit
r2_nakagawa <- r.squaredGLMM.merMod(fm)
r2_nakagawa %>% round(3) %>% knitr::kable()

contrast <- msstats_contrasts[1,]
lmerTestContrast(fm,contrast) %>% knitr::kable()

@

% re recursive clustering: necessary to resolve significant heterogenity which
% exists in large modules -- also improves recovery of biological signal.

\end{document}
