


<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
          (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-462338-14', 'auto');
  ga('require', 'GTM-TKM62DV');
  ga('send', 'pageview', {
    
  });


  var trackOutboundLink = function(url) {
    ga('send', 'event', 'outbound', 'click', url, {
       'transport': 'beacon',
       'hitCallback': function(){document.location = url;}
    });
  }
</script>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
    <meta property="og:title" content="DEqMS: vignettes/DEqMS-package-vignette.Rmd" />
    
      <meta name="description" content="DEqMS R Markdown vignettes">
      <meta property="og:description" content="DEqMS R Markdown vignettes"/>
    

    <link rel="icon" href="/favicon.ico">

    <link rel="amphtml" href="https://rdrr.io/a/bioc/DEqMS/f/vignettes/DEqMS-package-vignette.Rmd">
    <link rel="canonical" href="https://rdrr.io/bioc/DEqMS/f/vignettes/DEqMS-package-vignette.Rmd" />

    <link rel="search" type="application/opensearchdescription+xml" title="R Package Documentation" href="/opensearch.xml" />

    <!-- Hello from va1  -->

    <title>DEqMS: vignettes/DEqMS-package-vignette.Rmd</title>

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->

    
      
      
<link rel="stylesheet" href="/static/CACHE/css/e495a61847ad.css" type="text/css" />

    

    
  <script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Article",
  "headline": "DEqMS R Markdown vignettes",
  "datePublished": "2020-04-27T00:00:00",
  "dateModified": "2020-04-29T06:16:31.031499+00:00",
  "author": {
    "@type": "Person",
    "name": "Yafeng Zhu"
  },
  "description": "DEqMS R Markdown vignettes"
}
  </script>

  <style>
    
  </style>


    
      <script async src="//cdn.publift.com/fuse/tag/2/1306/fuse.js"></script>
    
  </head>

  <body>
    <div class="ui darkblue top fixed inverted menu" role="navigation" itemscope itemtype="http://www.schema.org/SiteNavigationElement" style="height: 40px; z-index: 1000;">
      <a class="ui header item " href="/" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'topbar-home', transport: 'beacon'});" >rdrr.io<!-- <small>R Package Documentation</small>--></a>
      <a class='ui item ' href="/find/" itemprop="url" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'topbar-find', transport: 'beacon'});" ><i class='search icon'></i><span itemprop="name">Find an R package</span></a>
      <a class='ui item ' href="/r/" itemprop="url" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'topbar-rdoc', transport: 'beacon'});" ><i class='file text outline icon'></i> <span itemprop="name">R language docs</span></a>
      <a class='ui item ' href="/snippets/" itemprop="url" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'topbar-snippets', transport: 'beacon'});" ><i class='play icon'></i> <span itemprop="name">Run R in your browser</span></a>
      <a class='ui item' href="https://rnotebook.io" target="_blank" rel="noopener"><i class='lab icon'></i> R Notebooks</a>

      <div class='right menu'>
        <form class='item' method='GET' action='/search'>
          <div class='ui right action input'>
            <input type='text' placeholder='packages, doc text, code...' size='24' name='q'>
            <button type="submit" class="ui green icon button"><i class='search icon'></i></button>
          </div>
        </form>
      </div>
    </div>

    
  




<div style='width: 280px; top: 24px; position: absolute;' class='ui vertical menu only-desktop bg-grey'>
  <a class='header  item' href='/bioc/DEqMS/' style='padding-bottom: 4px' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-pkg-head', transport: 'beacon'});" >
    <h3 class='ui header' style='margin-bottom: 4px'>
      DEqMS
      <div class='sub header'>a tool to perform statistical analysis of differential protein expression for quantitative proteomics data.</div>
    </h3>
    <small style='padding: 0 0 16px 0px' class='fakelink'>Package index</small>
  </a>

  <form class='item' method='GET' action='/search'>
    <div class='sub header' style='margin-bottom: 4px'>Search the DEqMS package</div>
    <div class='ui action input' style='padding-right: 32px'>
      <input type='hidden' name='package' value='DEqMS'>
      <input type='hidden' name='repo' value='bioc'>
      <input type='text' placeholder='' name='q'>
      <button type="submit" class="ui green icon button">
        <i class="search icon"></i>
      </button>
    </div>
  </form>

  
    <div class='header item' style='padding-bottom: 7px'>Vignettes</div>
    <small>
      <ul class='fakelist'>
        
          <li>
            <a href='/bioc/DEqMS/f/README.md' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-imp', transport: 'beacon'});" >
              README.md
              
            </a>
          </li>
        
          <li>
            <a href='/bioc/DEqMS/f/vignettes/DEqMS-package-vignette.Rmd' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-imp', transport: 'beacon'});" >
              DEqMS R Markdown vignettes
              
            </a>
          </li>
        
      </ul>
    </small>
  

  
<div class='ui floating dropdown item finder '>
  <b><a href='/bioc/DEqMS/api/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-index', transport: 'beacon'});" >Functions</a></b> <div class='ui blue label'>20</div>
  <i class='caret right icon'></i>
  
  
  
</div>

  
<div class='ui floating dropdown item finder '>
  <b><a href='/bioc/DEqMS/f/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-index', transport: 'beacon'});" >Source code</a></b> <div class='ui blue label'>4</div>
  <i class='caret right icon'></i>
  
  
  
</div>

  
<div class='ui floating dropdown item finder '>
  <b><a href='/bioc/DEqMS/man/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-index', transport: 'beacon'});" >Man pages</a></b> <div class='ui blue label'>11</div>
  <i class='caret right icon'></i>
  
    <small>
      <ul style='list-style-type: none; margin: 12px auto 0; line-height: 2.0; padding-left: 0px; padding-bottom: 8px;'>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/equalMedianNormalization.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>equalMedianNormalization: </b>normalize to have equal medians in all samples</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/farmsSummary.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>farmsSummary: </b>summarize peptide/PSM intensity into protein level relative...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/medianSummary.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>medianSummary: </b>summarize peptide/PSM intensity into protein level relative...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/medianSweeping.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>medianSweeping: </b>summarize peptide/PSM intensity into protein level relative...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/medpolishSummary.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>medpolishSummary: </b>summarize peptide/PSM intensity into protein level relative...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/outputResult.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>outputResult: </b>output the DEqMS analysis results in a data frame</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/peptideProfilePlot.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>peptideProfilePlot: </b>plot log2 intensities of all peptides for one gene in...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/Residualplot.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>Residualplot: </b>plot the residuals against the number of quantified...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/spectraCounteBayes.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>spectraCounteBayes: </b>Peptide/Spectra Count Based Empirical Bayes Statistics for...</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/VarianceBoxplot.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>VarianceBoxplot: </b>generate a boxplot of the variance</a></li>
        
          <li style='white-space: nowrap; text-overflow: clip; overflow: hidden;'><a href='/bioc/DEqMS/man/VarianceScatterplot.html' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-finder-obj', transport: 'beacon'});" ><b>VarianceScatterplot: </b>generate a scatter plot of the variance</a></li>
        
        <li style='padding-top: 4px; padding-bottom: 0;'><a href='/bioc/DEqMS/man/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'sidebar-bottom-browse-all', transport: 'beacon'});" ><b>Browse all...</b></a></li>
      </ul>
    </small>
  
  
  
</div>


  

  
</div>



  <div class='desktop-pad' id='body-content'>
    <div class='ui fluid container bc-row'>
      <div class='ui breadcrumb' itemscope itemtype="http://schema.org/BreadcrumbList">
        <a class='section' href="/">Home</a>

        <div class='divider'> / </div>

        <span itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem">
          <a class='section' itemscope itemtype="http://schema.org/Thing" itemprop="item" id="https://rdrr.io/all/bioc/" href="/all/bioc/">
            <span itemprop="name">Bioconductor</span>
          </a>
          <meta itemprop="position" content="1" />
        </span>

        <div class='divider'> / </div>

        <span itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem">
          <a class='section' itemscope itemtype="http://schema.org/Thing" itemprop="item" id="https://rdrr.io/bioc/DEqMS/" href="/bioc/DEqMS/">
            <span itemprop="name">DEqMS</span>
          </a>
          <meta itemprop="position" content="2" />
        </span>

        <div class='divider'> / </div>

        <span itemprop="itemListElement" itemscope itemtype="http://schema.org/ListItem" class="active section">
          <a class='active section' itemscope itemtype="http://schema.org/Thing" itemprop="item" href="https://rdrr.io/bioc/DEqMS/f/vignettes/DEqMS-package-vignette.Rmd">
            <span itemprop="name">DEqMS R Markdown vignettes</span>
          </a>
          <meta itemprop="position" content="3" /></li>
        </span>
      </div>
    </div>

    <div class='ui container' style='padding: 0px 8px'>
      <div class='only-desktop' style='float: right; width: 300px; height: 600px; margin-left: 16px;'></div>

      <h1 class='ui block header fit-content'>DEqMS R Markdown vignettes
        <div class='sub header'>In <a href='/bioc/DEqMS/'>DEqMS: a tool to perform statistical analysis of differential protein expression for quantitative proteomics data.</a>
      </h1>
      <h1>Overview of DEqMS</h1>
<p><code>DEqMS</code> builds on top of <code>Limma</code>, a widely-used R package for microarray data 
analysis (Smyth G. et al 2004), and improves it with proteomics data specific 
properties, accounting for variance dependence on the number of quantified 
peptides or PSMs for statistical testing of differential protein expression.</p>
<p>Limma assumes a common prior variance for all proteinss, the function 
<code>spectraCounteBayes</code> in DEqMS package estimate prior variance
for proteins quantified by different number of PSMs.</p>
<p>A documentation of all R functions available in DEqMS is detailed in the PDF 
reference manual on the DEqMS Bioconductor page.</p>
<h1>Load the package</h1>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">library</span>(DEqMS)
</pre></div>


<h1>Quick start</h1>
<h2>Differential protein expression analysis with DEqMS using a protein table</h2>
<p>As an example, we analyzed a protemoics dataset (TMT10plex labelled) in which 
A431 cells (human epidermoid carcinoma cell line) were treated with three 
different miRNA mimics (Zhou Y. Et al Oncogene 2017). The raw MS data was 
searched with MS-GF+ (Kim et al Nat Communications 2016) and post processed 
with Percolator (Kall L. et al Nat Method 2007). A tabular text output of 
protein table filtered at 1% protein level FDR is used.</p>
<h3>Download and Read the input protein table</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>url <span style="color: #666666">&lt;-</span> <span style="color: #BA2121">&quot;ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2016/06/PXD004163/Yan_miR_Protein_table.flatprottable.txt&quot;</span>
<span style="color: #0000FF">download.file</span>(url, destfile <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;./miR_Proteintable.txt&quot;</span>,method <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;auto&quot;</span>)

df.prot <span style="color: #666666">=</span> <span style="color: #0000FF">read.table</span>(<span style="color: #BA2121">&quot;miR_Proteintable.txt&quot;</span>,stringsAsFactors <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">FALSE</span>,
                     header <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>, quote <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>, comment.char <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>,sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;\t&quot;</span>)
</pre></div>


<h3>Extract quant data columns for DEqMS</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic"># filter at 1% protein FDR and extract TMT quantifications</span>
TMT_columns <span style="color: #666666">=</span> <span style="color: #0000FF">seq</span>(<span style="color: #666666">15</span>,<span style="color: #666666">33</span>,<span style="color: #666666">2</span>)
dat <span style="color: #666666">=</span> df.prot[df.prot<span style="color: #666666">$</span>miR.FASP_q.value<span style="color: #666666">&lt;0.01</span>,TMT_columns]
<span style="color: #0000FF">rownames</span>(dat) <span style="color: #666666">=</span> df.prot[df.prot<span style="color: #666666">$</span>miR.FASP_q.value<span style="color: #666666">&lt;0.01</span>,]<span style="color: #666666">$</span>Protein.accession
<span style="color: #408080; font-style: italic"># The protein dataframe is a typical protein expression matrix structure</span>
<span style="color: #408080; font-style: italic"># Samples are in columns, proteins are in rows</span>
<span style="color: #408080; font-style: italic"># use unique protein IDs for rownames</span>
<span style="color: #408080; font-style: italic"># to view the whole data frame, use the command View(dat)</span>
</pre></div>


<p>If the protein table is relative abundance (ratios) or intensity values, 
Log2 transform the data. Systematic effects and variance components are 
usually assumed to be additive on log scale (Oberg AL. et al JPR 2008; 
Hill EG. et al JPR 2008).</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>dat.log <span style="color: #666666">=</span> <span style="color: #0000FF">log2</span>(dat)
<span style="color: #408080; font-style: italic">#remove rows with NAs</span>
dat.log <span style="color: #666666">=</span> <span style="color: #0000FF">na.omit</span>(dat.log)
</pre></div>


<p>Use boxplot to check if the samples have medians centered.
if not, do median centering.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">boxplot</span>(dat.log,las<span style="color: #666666">=2</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TMT10plex data PXD004163&quot;</span>)
<span style="color: #408080; font-style: italic"># Here the data is already median centered, we skip the following step. </span>
<span style="color: #408080; font-style: italic"># dat.log = equalMedianNormalization(dat.log)</span>
</pre></div>


<h3>Make design table.</h3>
<p>A design table is used to tell how samples are arranged in
different groups/classes. </p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic"># if there is only one factor, such as treatment. You can define a vector with</span>
<span style="color: #408080; font-style: italic"># the treatment group in the same order as samples in the protein table.</span>
cond <span style="color: #666666">=</span> <span style="color: #0000FF">as.factor</span>(<span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191&quot;</span>,<span style="color: #BA2121">&quot;miR372&quot;</span>,<span style="color: #BA2121">&quot;miR519&quot;</span>,<span style="color: #BA2121">&quot;ctrl&quot;</span>,
<span style="color: #BA2121">&quot;miR372&quot;</span>,<span style="color: #BA2121">&quot;miR519&quot;</span>,<span style="color: #BA2121">&quot;ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191&quot;</span>,<span style="color: #BA2121">&quot;miR372&quot;</span>))

<span style="color: #408080; font-style: italic"># The function model.matrix is used to generate the design matrix</span>
design <span style="color: #666666">=</span> <span style="color: #0000FF">model.matrix</span>(<span style="color: #666666">~0+</span>cond) <span style="color: #408080; font-style: italic"># 0 means no intercept for the linear model</span>
<span style="color: #0000FF">colnames</span>(design) <span style="color: #666666">=</span> <span style="color: #0000FF">gsub</span>(<span style="color: #BA2121">&quot;cond&quot;</span>,<span style="color: #BA2121">&quot;&quot;</span>,<span style="color: #0000FF">colnames</span>(design))
</pre></div>


<h3>Make contrasts</h3>
<p>In addition to the design, you need to define the contrast, which tells the 
model to compare the differences between specific groups.
Start with the Limma part.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic"># you can define one or multiple contrasts here</span>
x <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;miR372-ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR519-ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191-ctrl&quot;</span>,
       <span style="color: #BA2121">&quot;miR372-miR519&quot;</span>,<span style="color: #BA2121">&quot;miR372-miR191&quot;</span>,<span style="color: #BA2121">&quot;miR519-miR191&quot;</span>)
contrast <span style="color: #666666">=</span>  <span style="color: #0000FF">makeContrasts</span>(contrasts<span style="color: #666666">=</span>x,levels<span style="color: #666666">=</span>design)
fit1 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">lmFit</span>(dat.log, design)
fit2 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">contrasts.fit</span>(fit1,contrasts <span style="color: #666666">=</span> contrast)
fit3 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">eBayes</span>(fit2)
</pre></div>


<h3>DEqMS analysis</h3>
<p>The above shows Limma part, now we use the function <code>spectraCounteBayes</code>
in DEqMS to correct bias of variance estimate based on minimum number of
psms per protein used for quantification.We use the minimum number of PSMs 
used for quantification within and across experiments to model the relation
between variance and PSM count.(See original paper)</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic"># assign a extra variable `count` to fit3 object, telling how many PSMs are </span>
<span style="color: #408080; font-style: italic"># quantifed for each protein</span>
<span style="color: #0000FF">library</span>(matrixStats)
count_columns <span style="color: #666666">=</span> <span style="color: #0000FF">seq</span>(<span style="color: #666666">16</span>,<span style="color: #666666">34</span>,<span style="color: #666666">2</span>)
psm.count.table <span style="color: #666666">=</span> <span style="color: #0000FF">data.frame</span>(count <span style="color: #666666">=</span> <span style="color: #0000FF">rowMins</span>(
  <span style="color: #0000FF">as.matrix</span>(df.prot[,count_columns])), row.names <span style="color: #666666">=</span>  df.prot<span style="color: #666666">$</span>Protein.accession)
fit3<span style="color: #666666">$</span>count <span style="color: #666666">=</span> psm.count.table<span style="color: #0000FF">[rownames</span>(fit3<span style="color: #666666">$</span>coefficients),<span style="color: #BA2121">&quot;count&quot;</span>]
fit4 <span style="color: #666666">=</span> <span style="color: #0000FF">spectraCounteBayes</span>(fit3)
</pre></div>


<p>Outputs of <code>spectraCounteBayes</code>:<br />
object is augmented form of "fit" object from <code>eBayes</code> in Limma, with the 
additions being:<br />
<code>sca.t</code> - Spectra Count Adjusted posterior t-value<br />
<code>sca.p</code> - Spectra Count Adjusted posterior p-value<br />
<code>sca.dfprior</code> - DEqMS estimated prior degrees of freedom<br />
<code>sca.priorvar</code>- DEqMS estimated prior variance<br />
<code>sca.postvar</code> - DEqMS estimated posterior variance<br />
<code>model</code> - fitted model</p>
<h3>Visualize the fit curve - variance dependence on quantified PSM</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic"># n=30 limits the boxplot to show only proteins quantified by &lt;= 30 PSMs.</span>
<span style="color: #0000FF">VarianceBoxplot</span>(fit4,n<span style="color: #666666">=30</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TMT10plex dataset PXD004163&quot;</span>,xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;PSM count&quot;</span>)
<span style="color: #0000FF">VarianceScatterplot</span>(fit4,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TMT10plex dataset PXD004163&quot;</span>)
</pre></div>


<h3>Extract the results as a data frame and save it</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>DEqMS.results <span style="color: #666666">=</span> <span style="color: #0000FF">outputResult</span>(fit4,coef_col <span style="color: #666666">=</span> <span style="color: #666666">1</span>)
<span style="color: #408080; font-style: italic">#if you are not sure which coef_col refers to the specific contrast,type</span>
<span style="color: #0000FF">head</span>(fit4<span style="color: #666666">$</span>coefficients)

<span style="color: #408080; font-style: italic"># a quick look on the DEqMS results table</span>
<span style="color: #0000FF">head</span>(DEqMS.results)
<span style="color: #408080; font-style: italic"># Save it into a tabular text file</span>
<span style="color: #0000FF">write.table</span>(DEqMS.results,<span style="color: #BA2121">&quot;DEqMS.results.miR372-ctrl.txt&quot;</span>,sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;\t&quot;</span>,
            row.names <span style="color: #666666">=</span> <span style="color: #008000">F</span>,quote<span style="color: #666666">=</span><span style="color: #008000">F</span>)
</pre></div>


<p>Explaination of the columns in <code>DEqMS.results</code>:<br />
<code>logFC</code> - log2 fold change between two groups, Here it's log2(miR372/ctrl). <br />
<code>AveExpr</code> - the mean of the log2 ratios/intensities across all samples. Since
input matrix is log2 ratio values, it is the mean log2 ratios of all samples. <br />
<code>t</code> - Limma output t-statistics<br />
<code>P.Value</code>- Limma p-values <br />
<code>adj.P.Val</code> - BH method adjusted Limma p-values<br />
<code>B</code> - Limma B values<br />
<code>count</code> - PSM/peptide count values you assigned<br />
<code>sca.t</code> - DEqMS t-statistics<br />
<code>sca.P.Value</code> - DEqMS p-values<br />
<code>sca.adj.pval</code> - BH method adjusted DEqMS p-values  </p>
<h3>Make volcanoplot</h3>
<p>We recommend to plot p-values on y-axis instead of adjusted pvalue or FDR.<br />
Read about why <a href="https://support.bioconductor.org/p/98442/">here</a>.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">library</span>(ggrepel)
<span style="color: #408080; font-style: italic"># Use ggplot2 allows more flexibility in plotting</span>

DEqMS.results<span style="color: #666666">$</span>log.sca.pval <span style="color: #666666">=</span> <span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(DEqMS.results<span style="color: #666666">$</span>sca.P.Value)
<span style="color: #0000FF">ggplot</span>(DEqMS.results, <span style="color: #0000FF">aes</span>(x <span style="color: #666666">=</span> logFC, y <span style="color: #666666">=</span>log.sca.pval )) <span style="color: #666666">+</span> 
    <span style="color: #0000FF">geom_point</span>(size<span style="color: #666666">=0.5</span> )<span style="color: #666666">+</span>
    <span style="color: #0000FF">theme_bw</span>(base_size <span style="color: #666666">=</span> <span style="color: #666666">16</span>) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># change theme</span>
    <span style="color: #0000FF">xlab</span>(<span style="color: #0000FF">expression</span>(<span style="color: #BA2121">&quot;log2(miR372/ctrl)&quot;</span>)) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># x-axis label</span>
    <span style="color: #0000FF">ylab</span>(<span style="color: #0000FF">expression</span>(<span style="color: #BA2121">&quot; -log10(P-value)&quot;</span>)) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># y-axis label</span>
    <span style="color: #0000FF">geom_vline</span>(xintercept <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #666666">-1</span>,<span style="color: #666666">1</span>), colour <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;red&quot;</span>) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># Add fold change cutoffs</span>
    <span style="color: #0000FF">geom_hline</span>(yintercept <span style="color: #666666">=</span> <span style="color: #666666">3</span>, colour <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;red&quot;</span>) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># Add significance cutoffs</span>
    <span style="color: #0000FF">geom_vline</span>(xintercept <span style="color: #666666">=</span> <span style="color: #666666">0</span>, colour <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;black&quot;</span>) <span style="color: #666666">+</span> <span style="color: #408080; font-style: italic"># Add 0 lines</span>
    <span style="color: #0000FF">scale_colour_gradient</span>(low <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;black&quot;</span>, high <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;black&quot;</span>, guide <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">FALSE</span>)<span style="color: #666666">+</span>
    <span style="color: #0000FF">geom_text_repel</span>(data<span style="color: #666666">=</span><span style="color: #0000FF">subset</span>(DEqMS.results, <span style="color: #0000FF">abs</span>(logFC)<span style="color: #666666">&gt;1&amp;</span>log.sca.pval <span style="color: #666666">&gt;</span> <span style="color: #666666">3</span>),
                    <span style="color: #0000FF">aes</span>( logFC, log.sca.pval ,label<span style="color: #666666">=</span>gene)) <span style="color: #408080; font-style: italic"># add gene label</span>
</pre></div>


<p>you can also use <code>volcanoplot</code> function from Limma. However, it uses <code>p.value</code> 
from Limma. If you want to plot <code>sca.pvalue</code> from DEqMS, you need to modify the 
<code>fit4</code> object.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>fit4<span style="color: #666666">$</span>p.value <span style="color: #666666">=</span> fit4<span style="color: #666666">$</span>sca.p
<span style="color: #408080; font-style: italic"># volcanoplot highlight top 20 proteins ranked by p-value here</span>
<span style="color: #0000FF">volcanoplot</span>(fit4,coef<span style="color: #666666">=1</span>, style <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;p-value&quot;</span>, highlight <span style="color: #666666">=</span> <span style="color: #666666">20</span>,
            names<span style="color: #666666">=</span><span style="color: #0000FF">rownames</span>(fit4<span style="color: #666666">$</span>coefficients))
</pre></div>


<h2>DEqMS analysis using MaxQuant outputs (label-free data)</h2>
<p>Here we analyze a published label-free benchmark dataset in which either 
10 or 30 µg of E. coli protein extract was spiked into human protein 
extracts (50 µg) in triplicates (Cox J et al MCP 2014). The data was searched 
by MaxQuant software and the output file "proteinGroups.txt" was used here.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>url2 <span style="color: #666666">&lt;-</span> <span style="color: #BA2121">&quot;ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2014/09/PXD000279/proteomebenchmark.zip&quot;</span>
<span style="color: #0000FF">download.file</span>(url2, destfile <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;./PXD000279.zip&quot;</span>,method <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;auto&quot;</span>)
<span style="color: #0000FF">unzip</span>(<span style="color: #BA2121">&quot;PXD000279.zip&quot;</span>)
</pre></div>


<h3>Read protein table as input and filter it</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>df.prot <span style="color: #666666">=</span> <span style="color: #0000FF">read.table</span>(<span style="color: #BA2121">&quot;proteinGroups.txt&quot;</span>,header<span style="color: #666666">=</span><span style="color: #008000">T</span>,sep<span style="color: #666666">=</span><span style="color: #BA2121">&quot;\t&quot;</span>,stringsAsFactors <span style="color: #666666">=</span> <span style="color: #008000">F</span>,
                        comment.char <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;&quot;</span>,quote <span style="color: #666666">=</span><span style="color: #BA2121">&quot;&quot;</span>)

<span style="color: #408080; font-style: italic"># remove decoy matches and matches to contaminant</span>
df.prot <span style="color: #666666">=</span> df.prot[<span style="color: #666666">!</span>df.prot<span style="color: #666666">$</span>Reverse<span style="color: #666666">==</span><span style="color: #BA2121">&quot;+&quot;</span>,]
df.prot <span style="color: #666666">=</span> df.prot[<span style="color: #666666">!</span>df.prot<span style="color: #666666">$</span>Contaminant<span style="color: #666666">==</span><span style="color: #BA2121">&quot;+&quot;</span>,]

<span style="color: #408080; font-style: italic"># Extract columns of LFQ intensites</span>
df.LFQ <span style="color: #666666">=</span> df.prot[,<span style="color: #666666">89:94</span>]
df.LFQ[df.LFQ<span style="color: #666666">==0</span>] <span style="color: #666666">&lt;-</span> <span style="color: #008000; font-weight: bold">NA</span>

<span style="color: #0000FF">rownames</span>(df.LFQ) <span style="color: #666666">=</span> df.prot<span style="color: #666666">$</span>Majority.protein.IDs
df.LFQ<span style="color: #666666">$</span>na_count_H <span style="color: #666666">=</span> <span style="color: #0000FF">apply</span>(df.LFQ,<span style="color: #666666">1</span>,<span style="color: #0000FF">function</span>(x) <span style="color: #0000FF">sum</span>(<span style="color: #0000FF">is.na</span>(x[1<span style="color: #666666">:3</span>])))
df.LFQ<span style="color: #666666">$</span>na_count_L <span style="color: #666666">=</span> <span style="color: #0000FF">apply</span>(df.LFQ,<span style="color: #666666">1</span>,<span style="color: #0000FF">function</span>(x) <span style="color: #0000FF">sum</span>(<span style="color: #0000FF">is.na</span>(x[4<span style="color: #666666">:6</span>])))
<span style="color: #408080; font-style: italic"># Filter protein table. DEqMS require minimum two values for each group.</span>
df.LFQ.filter <span style="color: #666666">=</span> df.LFQ[df.LFQ<span style="color: #666666">$</span>na_count_H<span style="color: #666666">&lt;2</span> <span style="color: #666666">&amp;</span> df.LFQ<span style="color: #666666">$</span>na_count_L<span style="color: #666666">&lt;2</span>,<span style="color: #666666">1:6</span>]
</pre></div>


<h3>Make a data frame of unique peptide count per protein</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">library</span>(matrixStats)
<span style="color: #408080; font-style: italic"># we use minimum peptide count among six samples</span>
<span style="color: #408080; font-style: italic"># count unique+razor peptides used for quantification</span>
pep.count.table <span style="color: #666666">=</span> <span style="color: #0000FF">data.frame</span>(count <span style="color: #666666">=</span> <span style="color: #0000FF">rowMins</span>(<span style="color: #0000FF">as.matrix</span>(df.prot[,<span style="color: #666666">19:24</span>])),
                             row.names <span style="color: #666666">=</span> df.prot<span style="color: #666666">$</span>Majority.protein.IDs)
<span style="color: #408080; font-style: italic"># Minimum peptide count of some proteins can be 0</span>
<span style="color: #408080; font-style: italic"># add pseudocount 1 to all proteins</span>
pep.count.table<span style="color: #666666">$</span>count <span style="color: #666666">=</span> pep.count.table<span style="color: #666666">$</span>count<span style="color: #666666">+1</span>
</pre></div>


<h3>DEqMS analysis on LFQ data</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>protein.matrix <span style="color: #666666">=</span> <span style="color: #0000FF">log2</span>(<span style="color: #0000FF">as.matrix</span>(df.LFQ.filter))

class <span style="color: #666666">=</span> <span style="color: #0000FF">as.factor</span>(<span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;H&quot;</span>,<span style="color: #BA2121">&quot;H&quot;</span>,<span style="color: #BA2121">&quot;H&quot;</span>,<span style="color: #BA2121">&quot;L&quot;</span>,<span style="color: #BA2121">&quot;L&quot;</span>,<span style="color: #BA2121">&quot;L&quot;</span>))
design <span style="color: #666666">=</span> <span style="color: #0000FF">model.matrix</span>(<span style="color: #666666">~0+</span>class) <span style="color: #408080; font-style: italic"># fitting without intercept</span>

fit1 <span style="color: #666666">=</span> <span style="color: #0000FF">lmFit</span>(protein.matrix,design <span style="color: #666666">=</span> design)
cont <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">makeContrasts</span>(classH<span style="color: #666666">-</span>classL, levels <span style="color: #666666">=</span> design)
fit2 <span style="color: #666666">=</span> <span style="color: #0000FF">contrasts.fit</span>(fit1,contrasts <span style="color: #666666">=</span> cont)
fit3 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">eBayes</span>(fit2)

fit3<span style="color: #666666">$</span>count <span style="color: #666666">=</span> pep.count.table<span style="color: #0000FF">[rownames</span>(fit3<span style="color: #666666">$</span>coefficients),<span style="color: #BA2121">&quot;count&quot;</span>]

<span style="color: #408080; font-style: italic">#check the values in the vector fit3$count</span>
<span style="color: #408080; font-style: italic">#if min(fit3$count) return NA or 0, you should troubleshoot the error first</span>
<span style="color: #0000FF">min</span>(fit3<span style="color: #666666">$</span>count)

fit4 <span style="color: #666666">=</span> <span style="color: #0000FF">spectraCounteBayes</span>(fit3)
</pre></div>


<h3>Visualize the fit curve</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">VarianceBoxplot</span>(fit4, n<span style="color: #666666">=20</span>, main <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;Label-free dataset PXD000279&quot;</span>,
                xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;peptide count + 1&quot;</span>)
</pre></div>


<h3>Extract outputs from DEqMS</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>DEqMS.results <span style="color: #666666">=</span> <span style="color: #0000FF">outputResult</span>(fit4,coef_col <span style="color: #666666">=</span> <span style="color: #666666">1</span>)
<span style="color: #408080; font-style: italic"># Add Gene names to the data frame</span>
<span style="color: #0000FF">rownames</span>(df.prot) <span style="color: #666666">=</span> df.prot<span style="color: #666666">$</span>Majority.protein.IDs
DEqMS.results<span style="color: #666666">$</span>Gene.name <span style="color: #666666">=</span> df.prot[DEqMS.results<span style="color: #666666">$</span>gene,]<span style="color: #666666">$</span>Gene.names
<span style="color: #0000FF">head</span>(DEqMS.results)
<span style="color: #0000FF">write.table</span>(DEqMS.results,<span style="color: #BA2121">&quot;H-L.DEqMS.results.txt&quot;</span>,sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;\t&quot;</span>,
            row.names <span style="color: #666666">=</span> <span style="color: #008000">F</span>,quote<span style="color: #666666">=</span><span style="color: #008000">F</span>)
</pre></div>


<h2>DEqMS analysis using a PSM table (isobaric labelled data)</h2>
<p>If you want to try different methods to estimate protein abundance,you can 
start with a PSM table and use provided functions in DEqMS to summarize PSM 
quant data into protein quant data. Four different functions are included: <code>medianSweeping</code>,<code>medianSummary</code>,<code>medpolishSummary</code>,<code>farmsSummary</code>. 
Check PDF reference manual for detailed description.</p>
<h3>Read PSM table input</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #408080; font-style: italic">### retrieve example PSM dataset from ExperimentHub</span>
<span style="color: #0000FF">library</span>(ExperimentHub)
eh <span style="color: #666666">=</span> <span style="color: #0000FF">ExperimentHub</span>()
<span style="color: #0000FF">query</span>(eh, <span style="color: #BA2121">&quot;DEqMS&quot;</span>)
dat.psm <span style="color: #666666">=</span> eh[[<span style="color: #BA2121">&quot;EH1663&quot;</span>]]
</pre></div>


<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>dat.psm.log <span style="color: #666666">=</span> dat.psm
dat.psm.log[,<span style="color: #666666">3:12</span>] <span style="color: #666666">=</span>  <span style="color: #0000FF">log2</span>(dat.psm[,<span style="color: #666666">3:12</span>])
<span style="color: #0000FF">head</span>(dat.psm.log)
</pre></div>


<h3>Summarization and Normalization</h3>
<p>Here, median sweeping is used to summarize PSMs intensities to protein log2 
ratios. In this procedure, we substract the spectrum log2 intensity from the 
median log2 intensities of all samples. The relative abundance estimate for 
each protein is calculated as the median over all PSMs belonging to this 
protein.(Herbrich et al JPR 2012 and D'Angelo et al JPR 2016).<br />
Assume the log2 intensity of PSM <code>i</code> in sample <code>j</code> is $y_{i,j}$, its relative 
log2 intensity of PSM <code>i</code> in sample <code>j</code> is $y'<em>{i,j}$: 
$$y'</em>{i,j} = y_{i,j} - median_{j'\in ctrl}\ y_{i,j'} $$
Relative abundance of protein <code>k</code> in sample <code>j</code> $Y_{k,j}$ is calculated as:
$$Y_{k,j} = median_{i\in protein\ k}\ y'_{i,j} $$</p>
<p>Correction for differences in amounts of material loaded in the channels 
is then done by subtracting the channel median from the relative abundance 
(log2 ratio), centering all channels to have median log2 value of zero.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>dat.gene.nm <span style="color: #666666">=</span> <span style="color: #0000FF">medianSweeping</span>(dat.psm.log,group_col <span style="color: #666666">=</span> <span style="color: #666666">2</span>)
<span style="color: #0000FF">boxplot</span>(dat.gene.nm,las<span style="color: #666666">=2</span>,ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;log2 ratio&quot;</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TMT10plex dataset PXD004163&quot;</span>)
</pre></div>


<h3>DEqMS analysis</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>gene.matrix <span style="color: #666666">=</span> <span style="color: #0000FF">as.matrix</span>(dat.gene.nm)

<span style="color: #408080; font-style: italic"># make design table</span>
cond <span style="color: #666666">=</span> <span style="color: #0000FF">as.factor</span>(<span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191&quot;</span>,<span style="color: #BA2121">&quot;miR372&quot;</span>,<span style="color: #BA2121">&quot;miR519&quot;</span>,<span style="color: #BA2121">&quot;ctrl&quot;</span>,
<span style="color: #BA2121">&quot;miR372&quot;</span>,<span style="color: #BA2121">&quot;miR519&quot;</span>,<span style="color: #BA2121">&quot;ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191&quot;</span>,<span style="color: #BA2121">&quot;miR372&quot;</span>))
design <span style="color: #666666">=</span> <span style="color: #0000FF">model.matrix</span>(<span style="color: #666666">~0+</span>cond) 
<span style="color: #0000FF">colnames</span>(design) <span style="color: #666666">=</span> <span style="color: #0000FF">gsub</span>(<span style="color: #BA2121">&quot;cond&quot;</span>,<span style="color: #BA2121">&quot;&quot;</span>,<span style="color: #0000FF">colnames</span>(design))

<span style="color: #408080; font-style: italic">#limma part analysis</span>
fit1 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">lmFit</span>(gene.matrix,design)
x <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;miR372-ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR519-ctrl&quot;</span>,<span style="color: #BA2121">&quot;miR191-ctrl&quot;</span>)
contrast <span style="color: #666666">=</span>  <span style="color: #0000FF">makeContrasts</span>(contrasts<span style="color: #666666">=</span>x,levels<span style="color: #666666">=</span>design)
fit2 <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">eBayes</span>(<span style="color: #0000FF">contrasts.fit</span>(fit1,contrasts <span style="color: #666666">=</span> contrast))

<span style="color: #408080; font-style: italic">#DEqMS part analysis</span>
psm.count.table <span style="color: #666666">=</span> <span style="color: #0000FF">as.data.frame</span>(<span style="color: #0000FF">table</span>(dat.psm<span style="color: #666666">$</span>gene))
<span style="color: #0000FF">rownames</span>(psm.count.table) <span style="color: #666666">=</span> psm.count.table<span style="color: #666666">$</span>Var1

fit2<span style="color: #666666">$</span>count <span style="color: #666666">=</span> psm.count.table<span style="color: #0000FF">[rownames</span>(fit2<span style="color: #666666">$</span>coefficients),<span style="color: #666666">2</span>]
fit3 <span style="color: #666666">=</span> <span style="color: #0000FF">spectraCounteBayes</span>(fit2)
<span style="color: #408080; font-style: italic"># extract DEqMS results</span>
DEqMS.results <span style="color: #666666">=</span> <span style="color: #0000FF">outputResult</span>(fit3,coef_col <span style="color: #666666">=</span> <span style="color: #666666">1</span>) 
<span style="color: #0000FF">head</span>(DEqMS.results)
<span style="color: #0000FF">write.table</span>(DEqMS.results,<span style="color: #BA2121">&quot;DEqMS.results.miR372-ctrl.fromPSMtable.txt&quot;</span>,
            sep <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;\t&quot;</span>,row.names <span style="color: #666666">=</span> <span style="color: #008000">F</span>,quote<span style="color: #666666">=</span><span style="color: #008000">F</span>)
</pre></div>


<p>Generate variance ~ PMS count boxplot, check if the DEqMS correctly find the relation between prior variance and PSM count</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">VarianceBoxplot</span>(fit3,n<span style="color: #666666">=20</span>, xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;PSM count&quot;</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TMT10plex dataset PXD004163&quot;</span>)
</pre></div>


<h3>PSM/Peptide profile plot</h3>
<p>Only possible if you read a PSM or peptide table as input.
<code>peptideProfilePlot</code> function will plot log2 intensity of each PSM/peptide of 
the protein in the input table.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">peptideProfilePlot</span>(dat<span style="color: #666666">=</span>dat.psm.log,col<span style="color: #666666">=2</span>,gene<span style="color: #666666">=</span><span style="color: #BA2121">&quot;TGFBR2&quot;</span>)
<span style="color: #408080; font-style: italic"># col=2 is tell in which column of dat.psm.log to look for the gene</span>
</pre></div>


<h1>Comparing DEqMS to other methods</h1>
<p>The following steps are not required for get the results from DEqMS.
it is used to help users to understand the method better and the differences
to other methods. Here we use the TMT labelled data PXD004163 as an example.</p>
<h2>Compare the variance estimate in DEqMS and Limma</h2>
<h3>Prior variance comparison between DEqMS and Limma</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">VarianceScatterplot</span>(fit3, xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;log2(PSM count)&quot;</span>)
limma.prior <span style="color: #666666">=</span> fit3<span style="color: #666666">$</span>s2.prior
<span style="color: #0000FF">abline</span>(h <span style="color: #666666">=</span> <span style="color: #0000FF">log</span>(limma.prior),col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;green&quot;</span>,lwd<span style="color: #666666">=3</span> )
<span style="color: #0000FF">legend</span>(<span style="color: #BA2121">&quot;topright&quot;</span>,legend<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;DEqMS prior variance&quot;</span>,<span style="color: #BA2121">&quot;Limma prior variance&quot;</span>),
        col<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;red&quot;</span>,<span style="color: #BA2121">&quot;green&quot;</span>),lwd<span style="color: #666666">=3</span>)
</pre></div>


<h3>Residual plot for DEqMS and Limma</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>op <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">par</span>(mfrow<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">1</span>,<span style="color: #666666">2</span>), mar<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">4</span>,<span style="color: #666666">4</span>,<span style="color: #666666">4</span>,<span style="color: #666666">1</span>), oma<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">0.5</span>,<span style="color: #666666">0.5</span>,<span style="color: #666666">0.5</span>,<span style="color: #666666">0</span>))
<span style="color: #0000FF">Residualplot</span>(fit3,  xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;log2(PSM count)&quot;</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;DEqMS&quot;</span>)
x <span style="color: #666666">=</span> fit3<span style="color: #666666">$</span>count
y <span style="color: #666666">=</span> <span style="color: #0000FF">log</span>(limma.prior) <span style="color: #666666">-</span> <span style="color: #0000FF">log</span>(fit3<span style="color: #666666">$</span>sigma^2)
<span style="color: #0000FF">plot</span>(<span style="color: #0000FF">log2</span>(x),y,ylim<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">-6</span>,<span style="color: #666666">2</span>),ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Variance(estimated-observed)&quot;</span>, pch<span style="color: #666666">=20</span>, cex<span style="color: #666666">=0.5</span>,
     xlab <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;log2(PSMcount)&quot;</span>,main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Limma&quot;</span>)
</pre></div>


<h3>Posterior variance comparison between DEqMS and Limma</h3>
<p>The plot here shows posterior variance of proteins "shrink" toward the 
fitted value to different extent depending on PSM number. </p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">library</span>(LSD)
op <span style="color: #666666">&lt;-</span> <span style="color: #0000FF">par</span>(mfrow<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">1</span>,<span style="color: #666666">2</span>), mar<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">4</span>,<span style="color: #666666">4</span>,<span style="color: #666666">4</span>,<span style="color: #666666">1</span>), oma<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">0.5</span>,<span style="color: #666666">0.5</span>,<span style="color: #666666">0.5</span>,<span style="color: #666666">0</span>))
x <span style="color: #666666">=</span> fit3<span style="color: #666666">$</span>count
y <span style="color: #666666">=</span> fit3<span style="color: #666666">$</span>s2.post
<span style="color: #0000FF">heatscatter</span>(<span style="color: #0000FF">log2</span>(x),<span style="color: #0000FF">log</span>(y),pch<span style="color: #666666">=20</span>, xlab <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;log2(PSMcount)&quot;</span>, 
     ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;log(Variance)&quot;</span>,
     main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Posterior Variance in Limma&quot;</span>)

y <span style="color: #666666">=</span> fit3<span style="color: #666666">$</span>sca.postvar
<span style="color: #0000FF">heatscatter</span>(<span style="color: #0000FF">log2</span>(x),<span style="color: #0000FF">log</span>(y),pch<span style="color: #666666">=20</span>, xlab <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;log2(PSMcount)&quot;</span>,
     ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;log(Variance)&quot;</span>, 
     main<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Posterior Variance in DEqMS&quot;</span>)
</pre></div>


<h2>Compare p-values from DEqMS to ordinary t-test, ANOVA and Limma</h2>
<p>We first apply t.test to detect significant protein changes between ctrl 
samples and miR372 treated samples, both have three replicates.</p>
<h3>T-test analysis</h3>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>pval.372 <span style="color: #666666">=</span> <span style="color: #0000FF">apply</span>(dat.gene.nm, <span style="color: #666666">1</span>, <span style="color: #0000FF">function</span>(x) 
<span style="color: #0000FF">t.test</span>(<span style="color: #0000FF">as.numeric</span>(x<span style="color: #0000FF">[c</span>(<span style="color: #666666">1</span>,<span style="color: #666666">5</span>,<span style="color: #666666">8</span>)]), <span style="color: #0000FF">as.numeric</span>(x<span style="color: #0000FF">[c</span>(<span style="color: #666666">3</span>,<span style="color: #666666">6</span>,<span style="color: #666666">10</span>)]))<span style="color: #666666">$</span>p.value)

logFC.372 <span style="color: #666666">=</span> <span style="color: #0000FF">rowMeans</span>(dat.gene.nm[,<span style="color: #0000FF">c</span>(<span style="color: #666666">3</span>,<span style="color: #666666">6</span>,<span style="color: #666666">10</span>)])<span style="color: #666666">-</span><span style="color: #0000FF">rowMeans</span>(dat.gene.nm[,<span style="color: #0000FF">c</span>(<span style="color: #666666">1</span>,<span style="color: #666666">5</span>,<span style="color: #666666">8</span>)])
</pre></div>


<p>Generate a data.frame of t.test results, add PSM count values and order the 
table by p-value.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>ttest.results <span style="color: #666666">=</span> <span style="color: #0000FF">data.frame</span>(gene<span style="color: #666666">=</span><span style="color: #0000FF">rownames</span>(dat.gene.nm),
                    logFC<span style="color: #666666">=</span>logFC.372,P.Value <span style="color: #666666">=</span> pval.372, 
                    adj.pval <span style="color: #666666">=</span> <span style="color: #0000FF">p.adjust</span>(pval.372,method <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;BH&quot;</span>)) 

ttest.results<span style="color: #666666">$</span>PSMcount <span style="color: #666666">=</span> psm.count.table[ttest.results<span style="color: #666666">$</span>gene,<span style="color: #BA2121">&quot;count&quot;</span>]
ttest.results <span style="color: #666666">=</span> ttest.results<span style="color: #0000FF">[with</span>(ttest.results, <span style="color: #0000FF">order</span>(P.Value)), ]
<span style="color: #0000FF">head</span>(ttest.results)
</pre></div>


<h3>Anova analysis</h3>
<p>Anova analysis is equivalent to linear model analysis. The difference to 
Limma analysis is that estimated variance is not moderated using empirical 
bayesian approach as it is done in Limma.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>ord.t <span style="color: #666666">=</span> fit1<span style="color: #666666">$</span>coefficients[, <span style="color: #666666">1</span>]<span style="color: #666666">/</span>fit1<span style="color: #666666">$</span>sigma<span style="color: #666666">/</span>fit1<span style="color: #666666">$</span>stdev.unscaled[, <span style="color: #666666">1</span>]
ord.p <span style="color: #666666">=</span> <span style="color: #666666">2*</span><span style="color: #0000FF">pt</span>(<span style="color: #0000FF">abs</span>(ord.t), fit1<span style="color: #666666">$</span>df.residual, lower.tail <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">FALSE</span>)
ord.q <span style="color: #666666">=</span> <span style="color: #0000FF">p.adjust</span>(ord.p,method <span style="color: #666666">=</span> <span style="color: #BA2121">&quot;BH&quot;</span>)
anova.results <span style="color: #666666">=</span> <span style="color: #0000FF">data.frame</span>(gene<span style="color: #666666">=</span><span style="color: #0000FF">names</span>(fit1<span style="color: #666666">$</span>sigma),
                            logFC<span style="color: #666666">=</span>fit1<span style="color: #666666">$</span>coefficients[,<span style="color: #666666">1</span>],
                            t<span style="color: #666666">=</span>ord.t, 
                            P.Value<span style="color: #666666">=</span>ord.p, 
                            adj.P.Val <span style="color: #666666">=</span> ord.q)

anova.results<span style="color: #666666">$</span>PSMcount <span style="color: #666666">=</span> psm.count.table[anova.results<span style="color: #666666">$</span>gene,<span style="color: #BA2121">&quot;count&quot;</span>]
anova.results <span style="color: #666666">=</span> anova.results<span style="color: #0000FF">[with</span>(anova.results,<span style="color: #0000FF">order</span>(P.Value)),]

<span style="color: #0000FF">head</span>(anova.results)
</pre></div>


<h3>Limma</h3>
<p>Extract limma results using <code>topTable</code> function,  <code>coef = 1</code> allows you to 
extract the specific contrast (miR372-ctrl), option <code>n= Inf</code> output 
all rows. </p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span>limma.results <span style="color: #666666">=</span> <span style="color: #0000FF">topTable</span>(fit2,coef <span style="color: #666666">=</span> <span style="color: #666666">1</span>,n<span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">Inf</span>)
limma.results<span style="color: #666666">$</span>gene <span style="color: #666666">=</span> <span style="color: #0000FF">rownames</span>(limma.results)
<span style="color: #408080; font-style: italic">#Add PSM count values in the data frame</span>
limma.results<span style="color: #666666">$</span>PSMcount <span style="color: #666666">=</span> psm.count.table[limma.results<span style="color: #666666">$</span>gene,<span style="color: #BA2121">&quot;count&quot;</span>]

<span style="color: #0000FF">head</span>(limma.results)
</pre></div>


<h3>Visualize the distribution of p-values by different analysis</h3>
<p>plotting all proteins ranked by p-values.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">plot</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(limma.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>), 
    type<span style="color: #666666">=</span><span style="color: #BA2121">&quot;l&quot;</span>,lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>, ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;-log10(p-value)&quot;</span>,ylim <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #666666">0</span>,<span style="color: #666666">10</span>),
    xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Proteins ranked by p-values&quot;</span>,
    col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;purple&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(DEqMS.results<span style="color: #666666">$</span>sca.P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>), 
        lty<span style="color: #666666">=1</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;red&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(anova.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>), 
        lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;blue&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(ttest.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>), 
        lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;orange&quot;</span>)
<span style="color: #0000FF">legend</span>(<span style="color: #BA2121">&quot;topright&quot;</span>,legend <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;Limma&quot;</span>,<span style="color: #BA2121">&quot;DEqMS&quot;</span>,<span style="color: #BA2121">&quot;Anova&quot;</span>,<span style="color: #BA2121">&quot;t.test&quot;</span>),
        col <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;purple&quot;</span>,<span style="color: #BA2121">&quot;red&quot;</span>,<span style="color: #BA2121">&quot;blue&quot;</span>,<span style="color: #BA2121">&quot;orange&quot;</span>),lty<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">2</span>,<span style="color: #666666">1</span>,<span style="color: #666666">2</span>,<span style="color: #666666">2</span>),lwd<span style="color: #666666">=2</span>)
</pre></div>


<p>plotting top 500 proteins ranked by p-values.</p>
<div class="codehilite" style="background: #f8f8f8"><pre style="line-height: 125%"><span></span><span style="color: #0000FF">plot</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(limma.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>)[1<span style="color: #666666">:500</span>], 
    type<span style="color: #666666">=</span><span style="color: #BA2121">&quot;l&quot;</span>,lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>, ylab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;-log10(p-value)&quot;</span>, ylim <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #666666">2</span>,<span style="color: #666666">10</span>),
    xlab<span style="color: #666666">=</span><span style="color: #BA2121">&quot;Proteins ranked by p-values&quot;</span>,
    col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;purple&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(DEqMS.results<span style="color: #666666">$</span>sca.P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>)[1<span style="color: #666666">:500</span>], 
        lty<span style="color: #666666">=1</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;red&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(anova.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>)[1<span style="color: #666666">:500</span>], 
        lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;blue&quot;</span>)
<span style="color: #0000FF">lines</span>(<span style="color: #0000FF">sort</span>(<span style="color: #666666">-</span><span style="color: #0000FF">log10</span>(ttest.results<span style="color: #666666">$</span>P.Value),decreasing <span style="color: #666666">=</span> <span style="color: #008000; font-weight: bold">TRUE</span>)[1<span style="color: #666666">:500</span>], 
        lty<span style="color: #666666">=2</span>,lwd<span style="color: #666666">=2</span>,col<span style="color: #666666">=</span><span style="color: #BA2121">&quot;orange&quot;</span>)
<span style="color: #0000FF">legend</span>(<span style="color: #BA2121">&quot;topright&quot;</span>,legend <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;Limma&quot;</span>,<span style="color: #BA2121">&quot;DEqMS&quot;</span>,<span style="color: #BA2121">&quot;Anova&quot;</span>,<span style="color: #BA2121">&quot;t.test&quot;</span>),
        col <span style="color: #666666">=</span> <span style="color: #0000FF">c</span>(<span style="color: #BA2121">&quot;purple&quot;</span>,<span style="color: #BA2121">&quot;red&quot;</span>,<span style="color: #BA2121">&quot;blue&quot;</span>,<span style="color: #BA2121">&quot;orange&quot;</span>),lty<span style="color: #666666">=</span><span style="color: #0000FF">c</span>(<span style="color: #666666">2</span>,<span style="color: #666666">1</span>,<span style="color: #666666">2</span>,<span style="color: #666666">2</span>),lwd<span style="color: #666666">=2</span>)
</pre></div>
      
      <br />
      <br />

      <div class='only-mobile' style='width: 320px; height: 100px; text-align: center;'></div>

      
  <h2 class='ui header'>Try the <a href="/bioc/DEqMS/">DEqMS</a> package in your browser</h2>

  <div class='ui form'>
    <div class='field'>
      <textarea class="mousetrap snip-input" id="snip" rows='10' cols='80' style='width: 100%; font-size: 13px; font-family: Menlo,Monaco,Consolas,"Courier New",monospace !important' data-tag='snrub'>library(DEqMS)

help(DEqMS)</textarea>
    </div>
  </div>

  <div class='ui container'>
    <div class='column'>
      <button class='ui huge green fluid button snip-run' data-tag='snrub' type="button" id="run">Run</button>
    </div>
    <div class='column'>
      <p><strong>Any scripts or data that you put into this service are public.</strong></p>
    </div>

    <div class='ui icon warning message snip-spinner hidden' data-tag='snrub'>
      <i class='notched circle loading icon'></i>
      <div class='content'>
        <div class='header snip-status' data-tag='snrub'>Nothing</div>
      </div>
    </div>

    <pre class='highlight hidden snip-output' data-tag='snrub'></pre>
    <div class='snip-images hidden' data-tag='snrub'></div>
  </div>


      <small><a href="/bioc/DEqMS/">DEqMS documentation</a> built on April 29, 2020, 6:16 a.m.</small>

    </div>

    

<div class="ui inverted darkblue vertical footer segment" style='margin-top: 16px; padding: 32px;'>
  <div class="ui center aligned container">
    <div class="ui stackable inverted divided three column centered grid">
      <div class="five wide column">
        <h4 class="ui inverted header">R Package Documentation</h4>
        <div class='ui inverted link list'>
          <a class='item' href='/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-home', transport: 'beacon'});" >rdrr.io home</a>
          <a class='item' href='/r/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-rdoc', transport: 'beacon'});" >R language documentation</a>
          <a class='item' href='/snippets/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-snippets', transport: 'beacon'});" >Run R code online</a>
          <a class='item' href='https://rnotebook.io/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-notebooks', transport: 'beacon'});" >Create free R Jupyter Notebooks</a>
        </div>
      </div>
      <div class="five wide column">
        <h4 class="ui inverted header">Browse R Packages</h4>
        <div class='ui inverted link list'>
          <a class='item' href='/all/cran/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-browse-cran', transport: 'beacon'});" >CRAN packages</a>
          <a class='item' href='/all/bioc/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-browse-bioc', transport: 'beacon'});" >Bioconductor packages</a>
          <a class='item' href='/all/rforge/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-browse-rforge', transport: 'beacon'});" >R-Forge packages</a>
          <a class='item' href='/all/github/' onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-browse-github', transport: 'beacon'});" >GitHub packages</a>
        </div>
      </div>
      <div class="five wide column">
        <h4 class="ui inverted header">We want your feedback!</h4>
        <small>Note that we can't provide technical support on individual packages. You should contact the package authors for that.</small>
        <div class='ui inverted link list'>
          <a class='item' href="https://twitter.com/intent/tweet?screen_name=rdrrHQ" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-twitter', transport: 'beacon'});" >
            <div class='ui large icon label twitter-button-colour'>
              <i class='whiteish twitter icon'></i> Tweet to @rdrrHQ
            </div>
          </a>

          <a class='item' href="https://github.com/rdrr-io/rdrr-issues/issues" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-issues', transport: 'beacon'});" >
            <div class='ui large icon label github-button-colour'>
              <i class='whiteish github icon'></i> GitHub issue tracker
            </div>
          </a>

          <a class='item' href="mailto:ian@mutexlabs.com" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-email', transport: 'beacon'});" >
            <div class='ui teal large icon label'>
              <i class='whiteish mail outline icon'></i> ian@mutexlabs.com
            </div>
          </a>

          <a class='item' href="https://ianhowson.com" onclick="ga('send', 'event', {eventCategory: 'click', eventAction: 'footer-blog', transport: 'beacon'});" >
            <div class='ui inverted large image label'>
              <img class='ui avatar image' src='/static/images/ianhowson32.png'> <span class='whiteish'>Personal blog</span>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>

  
  <br />
  <div class='only-mobile' style='min-height: 120px'>
    &nbsp;
  </div>
</div>

  </div>


    <!-- suggestions button -->
    <div style='position: fixed; bottom: 2%; right: 2%; z-index: 1000;'>
      <div class="ui raised segment surveyPopup" style='display:none'>
  <div class="ui large header">What can we improve?</div>

  <div class='content'>
    <div class="ui form">
      <div class="field">
        <button class='ui fluid button surveyReasonButton'>The page or its content looks wrong</button>
      </div>

      <div class="field">
        <button class='ui fluid button surveyReasonButton'>I can't find what I'm looking for</button>
      </div>

      <div class="field">
        <button class='ui fluid button surveyReasonButton'>I have a suggestion</button>
      </div>

      <div class="field">
        <button class='ui fluid button surveyReasonButton'>Other</button>
      </div>

      <div class="field">
        <label>Extra info (optional)</label>
        <textarea class='surveyText' rows='3' placeholder="Please enter more detail, if you like. Leave your email address if you'd like us to get in contact with you."></textarea>
      </div>

      <div class='ui error message surveyError' style='display: none'></div>

      <button class='ui large fluid green disabled button surveySubmitButton'>Submit</button>
    </div>
  </div>
</div>

      <button class='ui blue labeled icon button surveyButton only-desktop' style='display: none; float: right;'><i class="comment icon"></i> Improve this page</button>
      
    </div>

    

    
  


    <div class="ui modal snippetsModal">
  <div class="header">
    Embedding an R snippet on your website
  </div>
  <div class="content">
    <div class="description">
      <p>Add the following code to your website.</p>

      <p>
        <textarea class='codearea snippetEmbedCode' rows='5' style="font-family: Consolas,Monaco,'Andale Mono',monospace;">REMOVE THIS</textarea>
        <button class='ui blue button copyButton' data-clipboard-target='.snippetEmbedCode'>Copy to clipboard</button>
      </p>

      <p>For more information on customizing the embed code, read <a href='/snippets/embedding/'>Embedding Snippets</a>.</p>
    </div>
  </div>
  <div class="actions">
    <div class="ui button">Close</div>
  </div>
</div>

    
    <script type="text/javascript" src="/static/CACHE/js/73d0b6f91493.js"></script>

    
    <script type="text/javascript" src="/static/CACHE/js/5a4298830d2e.js"></script>

    
  

<script type="text/javascript">$(document).ready(function(){$('.snip-run').click(runClicked);var key='ctrl+enter';var txt=' (Ctrl-Enter)';if(navigator&&navigator.platform&&navigator.platform.startsWith&&navigator.platform.startsWith('Mac')){key='command+enter';txt=' (Cmd-Enter)';}
$('.snip-run').text('Run '+txt);Mousetrap.bind(key,function(e){if($('.snip-run').hasClass('disabled')){return;}
var faketarget=$('.snip-run')[0]
runClicked({currentTarget:faketarget});});});</script>



    
      
    

    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Open+Sans:400,400italic,600,600italic,800,800italic">
    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Oswald:400,300,700">
  </body>
</html>
